{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNAi Phenotypes Wormbase (WS264)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Author: Zachary Flamholz  \n",
    "Date: 06-2018  \n",
    "Database: https://wormbase.org/#012-34-5  \n",
    "Data: ftp://ftp.wormbase.org/pub/wormbase/releases/WS264/ONTOLOGY/rnai_phenotypes.WS264.wb  \n",
    "Companion file: http://ftp.ncbi.nlm.nih.gov/gene/DATA/GENE_INFO/Invertebrates/Caenorhabditis_elegans.gene_info.gz,  http://tazendra.caltech.edu/~azurebrd/cgi-bin/forms/generic.cgi?action=WpaXref  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Versions of Modules in Use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/json": {
       "Software versions": [
        {
         "module": "Python",
         "version": "2.7.15 64bit [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]"
        },
        {
         "module": "IPython",
         "version": "5.7.0"
        },
        {
         "module": "OS",
         "version": "Darwin 17.5.0 x86_64 i386 64bit"
        },
        {
         "module": "numpy",
         "version": "1.14.3"
        },
        {
         "module": "pandas",
         "version": "0.23.0"
        },
        {
         "module": "clustergrammer_widget",
         "version": "The 'clustergrammer_widget' distribution was not found and is required by the application"
        }
       ]
      },
      "text/html": [
       "<table><tr><th>Software</th><th>Version</th></tr><tr><td>Python</td><td>2.7.15 64bit [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]</td></tr><tr><td>IPython</td><td>5.7.0</td></tr><tr><td>OS</td><td>Darwin 17.5.0 x86_64 i386 64bit</td></tr><tr><td>numpy</td><td>1.14.3</td></tr><tr><td>pandas</td><td>0.23.0</td></tr><tr><td>clustergrammer_widget</td><td>The 'clustergrammer_widget' distribution was not found and is required by the application</td></tr><tr><td colspan='2'>Tue Jul 03 10:18:42 2018 EDT</td></tr></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{|l|l|}\\hline\n",
       "{\\bf Software} & {\\bf Version} \\\\ \\hline\\hline\n",
       "Python & 2.7.15 64bit [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE\\_401/final)] \\\\ \\hline\n",
       "IPython & 5.7.0 \\\\ \\hline\n",
       "OS & Darwin 17.5.0 x86\\_64 i386 64bit \\\\ \\hline\n",
       "numpy & 1.14.3 \\\\ \\hline\n",
       "pandas & 0.23.0 \\\\ \\hline\n",
       "clustergrammer_widget & The 'clustergrammer\\_widget' distribution was not found and is required by the application \\\\ \\hline\n",
       "\\hline \\multicolumn{2}{|l|}{Tue Jul 03 10:18:42 2018 EDT} \\\\ \\hline\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "Software versions\n",
       "Python 2.7.15 64bit [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]\n",
       "IPython 5.7.0\n",
       "OS Darwin 17.5.0 x86_64 i386 64bit\n",
       "numpy 1.14.3\n",
       "pandas 0.23.0\n",
       "clustergrammer_widget The 'clustergrammer_widget' distribution was not found and is required by the application\n",
       "Tue Jul 03 10:18:42 2018 EDT"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext version_information\n",
    "%version_information numpy, pandas, clustergrammer_widget "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import datetime\n",
    "import scipy.stats as stat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"input/rnai_phenotypes.WS264_WormBase.txt\", sep=\"\\t\", header=None)\n",
    "df.columns = [\"WB gene ID\", \"Gene symbol\", \"phenotype\", \"WB phenotype ID\", \"WB source ID\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WB gene ID</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>WB phenotype ID</th>\n",
       "      <th>WB source ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00001908</td>\n",
       "      <td>F17E9.9</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025129|WBPaper00006395 WBRNAi00025631|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00001908</td>\n",
       "      <td>F17E9.9</td>\n",
       "      <td>avoids bacterial lawn</td>\n",
       "      <td>WBPhenotype:0000402</td>\n",
       "      <td>WBRNAi00095640|WBPaper00040984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00001908</td>\n",
       "      <td>F17E9.9</td>\n",
       "      <td>RAB-11 recycling endosome localization variant</td>\n",
       "      <td>WBPhenotype:0002107</td>\n",
       "      <td>WBRNAi00090830|WBPaper00041129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00001908</td>\n",
       "      <td>F17E9.9</td>\n",
       "      <td>body elongation defective</td>\n",
       "      <td>WBPhenotype:0000242</td>\n",
       "      <td>WBRNAi00082054|WBPaper00005085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00001908</td>\n",
       "      <td>F17E9.9</td>\n",
       "      <td>chromosome segregation variant</td>\n",
       "      <td>WBPhenotype:0000773</td>\n",
       "      <td>WBRNAi00082059|WBPaper00005085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       WB gene ID Gene symbol                                       phenotype  \\\n",
       "0  WBGene00001908     F17E9.9                              locomotion variant   \n",
       "1  WBGene00001908     F17E9.9                           avoids bacterial lawn   \n",
       "2  WBGene00001908     F17E9.9  RAB-11 recycling endosome localization variant   \n",
       "3  WBGene00001908     F17E9.9                       body elongation defective   \n",
       "4  WBGene00001908     F17E9.9                  chromosome segregation variant   \n",
       "\n",
       "       WB phenotype ID                                       WB source ID  \n",
       "0  WBPhenotype:0000643  WBRNAi00025129|WBPaper00006395 WBRNAi00025631|...  \n",
       "1  WBPhenotype:0000402                     WBRNAi00095640|WBPaper00040984  \n",
       "2  WBPhenotype:0002107                     WBRNAi00090830|WBPaper00041129  \n",
       "3  WBPhenotype:0000242                     WBRNAi00082054|WBPaper00005085  \n",
       "4  WBPhenotype:0000773                     WBRNAi00082059|WBPaper00005085  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54540, 5)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1219"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df[\"phenotype\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8076"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df[\"Gene symbol\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_location_variant = df.loc[df[\"phenotype\"] == \"locomotion variant\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "WB gene ID                                            WBGene00001908\n",
       "Gene symbol                                                  F17E9.9\n",
       "phenotype                                         locomotion variant\n",
       "WB phenotype ID                                  WBPhenotype:0000643\n",
       "WB source ID       WBRNAi00025129|WBPaper00006395 WBRNAi00025631|...\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_location_variant.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'WBGene00001908'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_location_variant.iloc[0][\"WB gene ID\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## analyze term set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_terms = lambda x: len(df.loc[df[\"phenotype\"] == x])\n",
    "func = np.vectorize(count_terms)\n",
    "numGenes_perTerms = func(df[\"phenotype\"].unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44.741591468416736"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(numGenes_perTerms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median(numGenes_perTerms)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## convert gene symbol column to general nomenclature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "cElegans_geneInfo = pd.read_csv(\"input/Caenorhabditis_elegans.gene_info-1\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6239</td>\n",
       "      <td>171590</td>\n",
       "      <td>homt-1</td>\n",
       "      <td>CELE_Y74C9A.3</td>\n",
       "      <td>-</td>\n",
       "      <td>WormBase:WBGene00022277</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>Alpha N-terminal protein methyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Alpha N-terminal protein methyltransferase 1</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6239</td>\n",
       "      <td>171591</td>\n",
       "      <td>nlp-40</td>\n",
       "      <td>CELE_Y74C9A.2</td>\n",
       "      <td>-</td>\n",
       "      <td>WormBase:WBGene00022276</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>Peptide P4</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Peptide P4</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6239</td>\n",
       "      <td>171592</td>\n",
       "      <td>rcor-1</td>\n",
       "      <td>CELE_Y74C9A.4</td>\n",
       "      <td>-</td>\n",
       "      <td>WormBase:WBGene00022278</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>RCOR (REST CO-Repressor) homolog</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>RCOR (REST CO-Repressor) homolog</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6239</td>\n",
       "      <td>171593</td>\n",
       "      <td>sesn-1</td>\n",
       "      <td>CELE_Y74C9A.5</td>\n",
       "      <td>-</td>\n",
       "      <td>WormBase:WBGene00022279</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>Sestrin homolog</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Sestrin homolog</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6239</td>\n",
       "      <td>171594</td>\n",
       "      <td>pgs-1</td>\n",
       "      <td>CELE_Y48G1C.4</td>\n",
       "      <td>-</td>\n",
       "      <td>WormBase:WBGene00021677</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>CDP-diacylglycerol--glycerol-3-phosphate 3-pho...</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDP-diacylglycerol--glycerol-3-phosphate 3-pho...</td>\n",
       "      <td>20180406</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID  Symbol       LocusTag Synonyms                  dbXrefs  \\\n",
       "0     6239  171590  homt-1  CELE_Y74C9A.3        -  WormBase:WBGene00022277   \n",
       "1     6239  171591  nlp-40  CELE_Y74C9A.2        -  WormBase:WBGene00022276   \n",
       "2     6239  171592  rcor-1  CELE_Y74C9A.4        -  WormBase:WBGene00022278   \n",
       "3     6239  171593  sesn-1  CELE_Y74C9A.5        -  WormBase:WBGene00022279   \n",
       "4     6239  171594   pgs-1  CELE_Y48G1C.4        -  WormBase:WBGene00021677   \n",
       "\n",
       "  chromosome map_location                                        description  \\\n",
       "0          I            -       Alpha N-terminal protein methyltransferase 1   \n",
       "1          I            -                                         Peptide P4   \n",
       "2          I            -                   RCOR (REST CO-Repressor) homolog   \n",
       "3          I            -                                    Sestrin homolog   \n",
       "4          I            -  CDP-diacylglycerol--glycerol-3-phosphate 3-pho...   \n",
       "\n",
       "     type_of_gene Symbol_from_nomenclature_authority  \\\n",
       "0  protein-coding                                  -   \n",
       "1  protein-coding                                  -   \n",
       "2  protein-coding                                  -   \n",
       "3  protein-coding                                  -   \n",
       "4  protein-coding                                  -   \n",
       "\n",
       "  Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "0                                     -                   -   \n",
       "1                                     -                   -   \n",
       "2                                     -                   -   \n",
       "3                                     -                   -   \n",
       "4                                     -                   -   \n",
       "\n",
       "                                  Other_designations  Modification_date  \\\n",
       "0       Alpha N-terminal protein methyltransferase 1           20180606   \n",
       "1                                         Peptide P4           20180606   \n",
       "2                   RCOR (REST CO-Repressor) homolog           20180606   \n",
       "3                                    Sestrin homolog           20180606   \n",
       "4  CDP-diacylglycerol--glycerol-3-phosphate 3-pho...           20180406   \n",
       "\n",
       "  Feature_type  \n",
       "0            -  \n",
       "1            -  \n",
       "2            -  \n",
       "3            -  \n",
       "4            -  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cElegans_geneInfo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'WBGene00022277'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cElegans_geneInfo.iloc[0,5].split(\":\")[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6239</td>\n",
       "      <td>171590</td>\n",
       "      <td>homt-1</td>\n",
       "      <td>CELE_Y74C9A.3</td>\n",
       "      <td>-</td>\n",
       "      <td>WBGene00022277</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>Alpha N-terminal protein methyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Alpha N-terminal protein methyltransferase 1</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6239</td>\n",
       "      <td>171591</td>\n",
       "      <td>nlp-40</td>\n",
       "      <td>CELE_Y74C9A.2</td>\n",
       "      <td>-</td>\n",
       "      <td>WBGene00022276</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>Peptide P4</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Peptide P4</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6239</td>\n",
       "      <td>171592</td>\n",
       "      <td>rcor-1</td>\n",
       "      <td>CELE_Y74C9A.4</td>\n",
       "      <td>-</td>\n",
       "      <td>WBGene00022278</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>RCOR (REST CO-Repressor) homolog</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>RCOR (REST CO-Repressor) homolog</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6239</td>\n",
       "      <td>171593</td>\n",
       "      <td>sesn-1</td>\n",
       "      <td>CELE_Y74C9A.5</td>\n",
       "      <td>-</td>\n",
       "      <td>WBGene00022279</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>Sestrin homolog</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Sestrin homolog</td>\n",
       "      <td>20180606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6239</td>\n",
       "      <td>171594</td>\n",
       "      <td>pgs-1</td>\n",
       "      <td>CELE_Y48G1C.4</td>\n",
       "      <td>-</td>\n",
       "      <td>WBGene00021677</td>\n",
       "      <td>I</td>\n",
       "      <td>-</td>\n",
       "      <td>CDP-diacylglycerol--glycerol-3-phosphate 3-pho...</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDP-diacylglycerol--glycerol-3-phosphate 3-pho...</td>\n",
       "      <td>20180406</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID  Symbol       LocusTag Synonyms         dbXrefs chromosome  \\\n",
       "0     6239  171590  homt-1  CELE_Y74C9A.3        -  WBGene00022277          I   \n",
       "1     6239  171591  nlp-40  CELE_Y74C9A.2        -  WBGene00022276          I   \n",
       "2     6239  171592  rcor-1  CELE_Y74C9A.4        -  WBGene00022278          I   \n",
       "3     6239  171593  sesn-1  CELE_Y74C9A.5        -  WBGene00022279          I   \n",
       "4     6239  171594   pgs-1  CELE_Y48G1C.4        -  WBGene00021677          I   \n",
       "\n",
       "  map_location                                        description  \\\n",
       "0            -       Alpha N-terminal protein methyltransferase 1   \n",
       "1            -                                         Peptide P4   \n",
       "2            -                   RCOR (REST CO-Repressor) homolog   \n",
       "3            -                                    Sestrin homolog   \n",
       "4            -  CDP-diacylglycerol--glycerol-3-phosphate 3-pho...   \n",
       "\n",
       "     type_of_gene Symbol_from_nomenclature_authority  \\\n",
       "0  protein-coding                                  -   \n",
       "1  protein-coding                                  -   \n",
       "2  protein-coding                                  -   \n",
       "3  protein-coding                                  -   \n",
       "4  protein-coding                                  -   \n",
       "\n",
       "  Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "0                                     -                   -   \n",
       "1                                     -                   -   \n",
       "2                                     -                   -   \n",
       "3                                     -                   -   \n",
       "4                                     -                   -   \n",
       "\n",
       "                                  Other_designations  Modification_date  \\\n",
       "0       Alpha N-terminal protein methyltransferase 1           20180606   \n",
       "1                                         Peptide P4           20180606   \n",
       "2                   RCOR (REST CO-Repressor) homolog           20180606   \n",
       "3                                    Sestrin homolog           20180606   \n",
       "4  CDP-diacylglycerol--glycerol-3-phosphate 3-pho...           20180406   \n",
       "\n",
       "  Feature_type  \n",
       "0            -  \n",
       "1            -  \n",
       "2            -  \n",
       "3            -  \n",
       "4            -  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remove_WormBase = lambda x: x.replace(\"|\", \":\").split(\":\")[1] if \"WormBase\" in x else x\n",
    "## there are two rows that have muliple dbXrefs \n",
    "cElegans_geneInfo[\"dbXrefs\"] = cElegans_geneInfo[\"dbXrefs\"].apply(remove_WormBase)\n",
    "cElegans_geneInfo.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## testing how to convert names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(cElegans_geneInfo[\"dbXrefs\"][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(df.iloc[0,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17913    his-34\n",
       "Name: Symbol, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cElegans_geneInfo.loc[cElegans_geneInfo[\"dbXrefs\"] == df.iloc[0,0]][\"Symbol\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'WBGene00022277' in cElegans_geneInfo[\"dbXrefs\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WB gene ID</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>WB phenotype ID</th>\n",
       "      <th>WB source ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00001908</td>\n",
       "      <td>F17E9.9</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025129|WBPaper00006395 WBRNAi00025631|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>WBGene00007110</td>\n",
       "      <td>B0035.11</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00024350|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>WBGene00006707</td>\n",
       "      <td>R09B3.4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026070|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>WBGene00003808</td>\n",
       "      <td>T05A1.1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00075428|WBPaper00006132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>WBGene00016250</td>\n",
       "      <td>C30C11.4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00024683|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>WBGene00012735</td>\n",
       "      <td>Y40B1A.4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00004607|WBPaper00004402 WBRNAi00026682|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>WBGene00004951</td>\n",
       "      <td>K10B3.10</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008979|WBPaper00005654 WBRNAi00025934|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>WBGene00017678</td>\n",
       "      <td>F21F8.4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00078339|WBPaper00027214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>WBGene00019778</td>\n",
       "      <td>M57.2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009019|WBPaper00005654 WBRNAi00026001|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>WBGene00019881</td>\n",
       "      <td>R05D3.8</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026026|WBPaper00006395 WBRNAi00074594|...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         WB gene ID Gene symbol           phenotype      WB phenotype ID  \\\n",
       "0    WBGene00001908     F17E9.9  locomotion variant  WBPhenotype:0000643   \n",
       "57   WBGene00007110    B0035.11  locomotion variant  WBPhenotype:0000643   \n",
       "118  WBGene00006707     R09B3.4  locomotion variant  WBPhenotype:0000643   \n",
       "168  WBGene00003808     T05A1.1  locomotion variant  WBPhenotype:0000643   \n",
       "185  WBGene00016250    C30C11.4  locomotion variant  WBPhenotype:0000643   \n",
       "226  WBGene00012735    Y40B1A.4  locomotion variant  WBPhenotype:0000643   \n",
       "289  WBGene00004951    K10B3.10  locomotion variant  WBPhenotype:0000643   \n",
       "355  WBGene00017678     F21F8.4  locomotion variant  WBPhenotype:0000643   \n",
       "393  WBGene00019778       M57.2  locomotion variant  WBPhenotype:0000643   \n",
       "535  WBGene00019881     R05D3.8  locomotion variant  WBPhenotype:0000643   \n",
       "\n",
       "                                          WB source ID  \n",
       "0    WBRNAi00025129|WBPaper00006395 WBRNAi00025631|...  \n",
       "57                      WBRNAi00024350|WBPaper00006395  \n",
       "118                     WBRNAi00026070|WBPaper00006395  \n",
       "168                     WBRNAi00075428|WBPaper00006132  \n",
       "185                     WBRNAi00024683|WBPaper00006395  \n",
       "226  WBRNAi00004607|WBPaper00004402 WBRNAi00026682|...  \n",
       "289  WBRNAi00008979|WBPaper00005654 WBRNAi00025934|...  \n",
       "355                     WBRNAi00078339|WBPaper00027214  \n",
       "393  WBRNAi00009019|WBPaper00005654 WBRNAi00026001|...  \n",
       "535  WBRNAi00026026|WBPaper00006395 WBRNAi00074594|...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_location_variant.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(cElegans_geneInfo.loc[cElegans_geneInfo[\"dbXrefs\"] == 'WBGene00001908'][\"Symbol\"].tolist()[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "## if there is no WB gene ID in the reference table then the symbol becomes the WB gene ID from the phenotype list\n",
    "change_name = lambda x: cElegans_geneInfo.loc[cElegans_geneInfo[\"dbXrefs\"] == x][\"Symbol\"].tolist()[0] if x in cElegans_geneInfo[\"dbXrefs\"].values else x\n",
    "df_location_variant[\"Gene symbol\"] = df_location_variant[\"WB gene ID\"].apply(change_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WB gene ID</th>\n",
       "      <th>Gene symbol</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>WB phenotype ID</th>\n",
       "      <th>WB source ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00001908</td>\n",
       "      <td>his-34</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025129|WBPaper00006395 WBRNAi00025631|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>WBGene00007110</td>\n",
       "      <td>leo-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00024350|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>WBGene00006707</td>\n",
       "      <td>ubc-12</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026070|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>WBGene00003808</td>\n",
       "      <td>npr-2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00075428|WBPaper00006132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>WBGene00016250</td>\n",
       "      <td>hsp-110</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00024683|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>WBGene00012735</td>\n",
       "      <td>sptf-3</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00004607|WBPaper00004402 WBRNAi00026682|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>WBGene00004951</td>\n",
       "      <td>spc-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008979|WBPaper00005654 WBRNAi00025934|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>WBGene00017678</td>\n",
       "      <td>asp-12</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00078339|WBPaper00027214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>WBGene00019778</td>\n",
       "      <td>M57.2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009019|WBPaper00005654 WBRNAi00026001|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>WBGene00019881</td>\n",
       "      <td>R05D3.8</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026026|WBPaper00006395 WBRNAi00074594|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>WBGene00006698</td>\n",
       "      <td>uaf-2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00075324|WBPaper00029258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>WBGene00001694</td>\n",
       "      <td>grd-5</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00077019|WBPaper00026841 WBRNAi00077023|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>WBGene00006745</td>\n",
       "      <td>unc-5</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00089729|WBPaper00040284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>WBGene00003171</td>\n",
       "      <td>mec-7</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00075171|WBPaper00029258 WBRNAi00075173|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>WBGene00001701</td>\n",
       "      <td>grd-12</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00077024|WBPaper00026841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754</th>\n",
       "      <td>WBGene00001949</td>\n",
       "      <td>hlh-2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009007|WBPaper00005654 WBRNAi00077261|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>WBGene00016170</td>\n",
       "      <td>tmem-131</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008452|WBPaper00005654 WBRNAi00024656|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>WBGene00000105</td>\n",
       "      <td>alg-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008819|WBPaper00005654 WBRNAi00025553|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>WBGene00018819</td>\n",
       "      <td>rog-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00001045|WBPaper00004651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>889</th>\n",
       "      <td>WBGene00018027</td>\n",
       "      <td>F35A5.5</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00027318|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>WBGene00017241</td>\n",
       "      <td>pcyt-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025032|WBPaper00006395 WBRNAi00027260|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>979</th>\n",
       "      <td>WBGene00013168</td>\n",
       "      <td>arp-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009379|WBPaper00005654 WBRNAi00026815|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>WBGene00006382</td>\n",
       "      <td>taf-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026477|WBPaper00006395 WBRNAi00026478|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1070</th>\n",
       "      <td>WBGene00003939</td>\n",
       "      <td>pax-3</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008706|WBPaper00005654 WBRNAi00025285|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1111</th>\n",
       "      <td>WBGene00000161</td>\n",
       "      <td>apa-2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009191|WBPaper00005654 WBRNAi00026328|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1177</th>\n",
       "      <td>WBGene00000739</td>\n",
       "      <td>col-166</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026228|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1238</th>\n",
       "      <td>WBGene00004873</td>\n",
       "      <td>smc-3</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026718|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1442</th>\n",
       "      <td>WBGene00013343</td>\n",
       "      <td>prp-6</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026881|WBPaper00006395 WBRNAi00037628|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>WBGene00006516</td>\n",
       "      <td>vps-16</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00072897|WBPaper00029258 WBRNAi00072898|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1478</th>\n",
       "      <td>WBGene00001885</td>\n",
       "      <td>his-11</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025631|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2972</th>\n",
       "      <td>WBGene00018699</td>\n",
       "      <td>WBGene00018699</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00070598|WBPaper00004841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3013</th>\n",
       "      <td>WBGene00012704</td>\n",
       "      <td>mlt-3</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009314|WBPaper00005654 WBRNAi00026666|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3033</th>\n",
       "      <td>WBGene00010204</td>\n",
       "      <td>F57F5.1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025738|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3099</th>\n",
       "      <td>WBGene00007008</td>\n",
       "      <td>rfp-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026026|WBPaper00006395 WBRNAi00074594|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3130</th>\n",
       "      <td>WBGene00003925</td>\n",
       "      <td>pas-4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00024749|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3173</th>\n",
       "      <td>WBGene00001328</td>\n",
       "      <td>epi-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008971|WBPaper00005654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3251</th>\n",
       "      <td>WBGene00000741</td>\n",
       "      <td>col-168</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00026228|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3261</th>\n",
       "      <td>WBGene00004341</td>\n",
       "      <td>rfl-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00027275|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3279</th>\n",
       "      <td>WBGene00004829</td>\n",
       "      <td>sli-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00109167|WBPaper00046120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3285</th>\n",
       "      <td>WBGene00011758</td>\n",
       "      <td>T13H5.4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00075301|WBPaper00029258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>WBGene00010785</td>\n",
       "      <td>top-2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00078335|WBPaper00027214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365</th>\n",
       "      <td>WBGene00012261</td>\n",
       "      <td>lpr-3</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009253|WBPaper00005654 WBRNAi00026489|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3410</th>\n",
       "      <td>WBGene00001517</td>\n",
       "      <td>gar-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00075430|WBPaper00006132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3416</th>\n",
       "      <td>WBGene00001087</td>\n",
       "      <td>dpy-28</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009310|WBPaper00005654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3528</th>\n",
       "      <td>WBGene00003952</td>\n",
       "      <td>pbs-6</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008327|WBPaper00005654 WBRNAi00024451|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594</th>\n",
       "      <td>WBGene00001876</td>\n",
       "      <td>his-2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025512|WBPaper00006395 WBRNAi00026260|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3646</th>\n",
       "      <td>WBGene00000218</td>\n",
       "      <td>asp-5</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00078339|WBPaper00027214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3652</th>\n",
       "      <td>WBGene00008053</td>\n",
       "      <td>cdc-48.2</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00024790|WBPaper00006395 WBRNAi00078676|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3774</th>\n",
       "      <td>WBGene00001019</td>\n",
       "      <td>dnj-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00027137|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3778</th>\n",
       "      <td>WBGene00018270</td>\n",
       "      <td>F41C3.4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008776|WBPaper00005654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3792</th>\n",
       "      <td>WBGene00000931</td>\n",
       "      <td>dao-5</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00024619|WBPaper00006395 WBRNAi00024620|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3840</th>\n",
       "      <td>WBGene00004259</td>\n",
       "      <td>pyr-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008581|WBPaper00005654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3856</th>\n",
       "      <td>WBGene00001570</td>\n",
       "      <td>gei-13</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008892|WBPaper00005654 WBRNAi00025742|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3950</th>\n",
       "      <td>WBGene00003639</td>\n",
       "      <td>nhr-49</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025936|WBPaper00006395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4109</th>\n",
       "      <td>WBGene00006584</td>\n",
       "      <td>tni-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025481|WBPaper00006395 WBRNAi00027336|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>WBGene00006432</td>\n",
       "      <td>WBGene00006432</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025372|WBPaper00006395 WBRNAi00027316|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4171</th>\n",
       "      <td>WBGene00006829</td>\n",
       "      <td>unc-101</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025950|WBPaper00006395 WBRNAi00025951|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4189</th>\n",
       "      <td>WBGene00019154</td>\n",
       "      <td>glf-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00008915|WBPaper00005654 WBRNAi00025787|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4250</th>\n",
       "      <td>WBGene00185088</td>\n",
       "      <td>Y55F3BL.4</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00009393|WBPaper00005654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4271</th>\n",
       "      <td>WBGene00008775</td>\n",
       "      <td>mogs-1</td>\n",
       "      <td>locomotion variant</td>\n",
       "      <td>WBPhenotype:0000643</td>\n",
       "      <td>WBRNAi00025094|WBPaper00006395 WBRNAi00027277|...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          WB gene ID     Gene symbol           phenotype      WB phenotype ID  \\\n",
       "0     WBGene00001908          his-34  locomotion variant  WBPhenotype:0000643   \n",
       "57    WBGene00007110           leo-1  locomotion variant  WBPhenotype:0000643   \n",
       "118   WBGene00006707          ubc-12  locomotion variant  WBPhenotype:0000643   \n",
       "168   WBGene00003808           npr-2  locomotion variant  WBPhenotype:0000643   \n",
       "185   WBGene00016250         hsp-110  locomotion variant  WBPhenotype:0000643   \n",
       "226   WBGene00012735          sptf-3  locomotion variant  WBPhenotype:0000643   \n",
       "289   WBGene00004951           spc-1  locomotion variant  WBPhenotype:0000643   \n",
       "355   WBGene00017678          asp-12  locomotion variant  WBPhenotype:0000643   \n",
       "393   WBGene00019778           M57.2  locomotion variant  WBPhenotype:0000643   \n",
       "535   WBGene00019881         R05D3.8  locomotion variant  WBPhenotype:0000643   \n",
       "565   WBGene00006698           uaf-2  locomotion variant  WBPhenotype:0000643   \n",
       "603   WBGene00001694           grd-5  locomotion variant  WBPhenotype:0000643   \n",
       "638   WBGene00006745           unc-5  locomotion variant  WBPhenotype:0000643   \n",
       "654   WBGene00003171           mec-7  locomotion variant  WBPhenotype:0000643   \n",
       "705   WBGene00001701          grd-12  locomotion variant  WBPhenotype:0000643   \n",
       "754   WBGene00001949           hlh-2  locomotion variant  WBPhenotype:0000643   \n",
       "783   WBGene00016170        tmem-131  locomotion variant  WBPhenotype:0000643   \n",
       "823   WBGene00000105           alg-1  locomotion variant  WBPhenotype:0000643   \n",
       "863   WBGene00018819           rog-1  locomotion variant  WBPhenotype:0000643   \n",
       "889   WBGene00018027         F35A5.5  locomotion variant  WBPhenotype:0000643   \n",
       "908   WBGene00017241          pcyt-1  locomotion variant  WBPhenotype:0000643   \n",
       "979   WBGene00013168           arp-1  locomotion variant  WBPhenotype:0000643   \n",
       "1032  WBGene00006382           taf-1  locomotion variant  WBPhenotype:0000643   \n",
       "1070  WBGene00003939           pax-3  locomotion variant  WBPhenotype:0000643   \n",
       "1111  WBGene00000161           apa-2  locomotion variant  WBPhenotype:0000643   \n",
       "1177  WBGene00000739         col-166  locomotion variant  WBPhenotype:0000643   \n",
       "1238  WBGene00004873           smc-3  locomotion variant  WBPhenotype:0000643   \n",
       "1442  WBGene00013343           prp-6  locomotion variant  WBPhenotype:0000643   \n",
       "1462  WBGene00006516          vps-16  locomotion variant  WBPhenotype:0000643   \n",
       "1478  WBGene00001885          his-11  locomotion variant  WBPhenotype:0000643   \n",
       "...              ...             ...                 ...                  ...   \n",
       "2972  WBGene00018699  WBGene00018699  locomotion variant  WBPhenotype:0000643   \n",
       "3013  WBGene00012704           mlt-3  locomotion variant  WBPhenotype:0000643   \n",
       "3033  WBGene00010204         F57F5.1  locomotion variant  WBPhenotype:0000643   \n",
       "3099  WBGene00007008           rfp-1  locomotion variant  WBPhenotype:0000643   \n",
       "3130  WBGene00003925           pas-4  locomotion variant  WBPhenotype:0000643   \n",
       "3173  WBGene00001328           epi-1  locomotion variant  WBPhenotype:0000643   \n",
       "3251  WBGene00000741         col-168  locomotion variant  WBPhenotype:0000643   \n",
       "3261  WBGene00004341           rfl-1  locomotion variant  WBPhenotype:0000643   \n",
       "3279  WBGene00004829           sli-1  locomotion variant  WBPhenotype:0000643   \n",
       "3285  WBGene00011758         T13H5.4  locomotion variant  WBPhenotype:0000643   \n",
       "3318  WBGene00010785           top-2  locomotion variant  WBPhenotype:0000643   \n",
       "3365  WBGene00012261           lpr-3  locomotion variant  WBPhenotype:0000643   \n",
       "3410  WBGene00001517           gar-1  locomotion variant  WBPhenotype:0000643   \n",
       "3416  WBGene00001087          dpy-28  locomotion variant  WBPhenotype:0000643   \n",
       "3528  WBGene00003952           pbs-6  locomotion variant  WBPhenotype:0000643   \n",
       "3594  WBGene00001876           his-2  locomotion variant  WBPhenotype:0000643   \n",
       "3646  WBGene00000218           asp-5  locomotion variant  WBPhenotype:0000643   \n",
       "3652  WBGene00008053        cdc-48.2  locomotion variant  WBPhenotype:0000643   \n",
       "3774  WBGene00001019           dnj-1  locomotion variant  WBPhenotype:0000643   \n",
       "3778  WBGene00018270         F41C3.4  locomotion variant  WBPhenotype:0000643   \n",
       "3792  WBGene00000931           dao-5  locomotion variant  WBPhenotype:0000643   \n",
       "3840  WBGene00004259           pyr-1  locomotion variant  WBPhenotype:0000643   \n",
       "3856  WBGene00001570          gei-13  locomotion variant  WBPhenotype:0000643   \n",
       "3950  WBGene00003639          nhr-49  locomotion variant  WBPhenotype:0000643   \n",
       "4109  WBGene00006584           tni-1  locomotion variant  WBPhenotype:0000643   \n",
       "4126  WBGene00006432  WBGene00006432  locomotion variant  WBPhenotype:0000643   \n",
       "4171  WBGene00006829         unc-101  locomotion variant  WBPhenotype:0000643   \n",
       "4189  WBGene00019154           glf-1  locomotion variant  WBPhenotype:0000643   \n",
       "4250  WBGene00185088       Y55F3BL.4  locomotion variant  WBPhenotype:0000643   \n",
       "4271  WBGene00008775          mogs-1  locomotion variant  WBPhenotype:0000643   \n",
       "\n",
       "                                           WB source ID  \n",
       "0     WBRNAi00025129|WBPaper00006395 WBRNAi00025631|...  \n",
       "57                       WBRNAi00024350|WBPaper00006395  \n",
       "118                      WBRNAi00026070|WBPaper00006395  \n",
       "168                      WBRNAi00075428|WBPaper00006132  \n",
       "185                      WBRNAi00024683|WBPaper00006395  \n",
       "226   WBRNAi00004607|WBPaper00004402 WBRNAi00026682|...  \n",
       "289   WBRNAi00008979|WBPaper00005654 WBRNAi00025934|...  \n",
       "355                      WBRNAi00078339|WBPaper00027214  \n",
       "393   WBRNAi00009019|WBPaper00005654 WBRNAi00026001|...  \n",
       "535   WBRNAi00026026|WBPaper00006395 WBRNAi00074594|...  \n",
       "565                      WBRNAi00075324|WBPaper00029258  \n",
       "603   WBRNAi00077019|WBPaper00026841 WBRNAi00077023|...  \n",
       "638                      WBRNAi00089729|WBPaper00040284  \n",
       "654   WBRNAi00075171|WBPaper00029258 WBRNAi00075173|...  \n",
       "705                      WBRNAi00077024|WBPaper00026841  \n",
       "754   WBRNAi00009007|WBPaper00005654 WBRNAi00077261|...  \n",
       "783   WBRNAi00008452|WBPaper00005654 WBRNAi00024656|...  \n",
       "823   WBRNAi00008819|WBPaper00005654 WBRNAi00025553|...  \n",
       "863                      WBRNAi00001045|WBPaper00004651  \n",
       "889                      WBRNAi00027318|WBPaper00006395  \n",
       "908   WBRNAi00025032|WBPaper00006395 WBRNAi00027260|...  \n",
       "979   WBRNAi00009379|WBPaper00005654 WBRNAi00026815|...  \n",
       "1032  WBRNAi00026477|WBPaper00006395 WBRNAi00026478|...  \n",
       "1070  WBRNAi00008706|WBPaper00005654 WBRNAi00025285|...  \n",
       "1111  WBRNAi00009191|WBPaper00005654 WBRNAi00026328|...  \n",
       "1177                     WBRNAi00026228|WBPaper00006395  \n",
       "1238                     WBRNAi00026718|WBPaper00006395  \n",
       "1442  WBRNAi00026881|WBPaper00006395 WBRNAi00037628|...  \n",
       "1462  WBRNAi00072897|WBPaper00029258 WBRNAi00072898|...  \n",
       "1478                     WBRNAi00025631|WBPaper00006395  \n",
       "...                                                 ...  \n",
       "2972                     WBRNAi00070598|WBPaper00004841  \n",
       "3013  WBRNAi00009314|WBPaper00005654 WBRNAi00026666|...  \n",
       "3033                     WBRNAi00025738|WBPaper00006395  \n",
       "3099  WBRNAi00026026|WBPaper00006395 WBRNAi00074594|...  \n",
       "3130                     WBRNAi00024749|WBPaper00006395  \n",
       "3173                     WBRNAi00008971|WBPaper00005654  \n",
       "3251                     WBRNAi00026228|WBPaper00006395  \n",
       "3261                     WBRNAi00027275|WBPaper00006395  \n",
       "3279                     WBRNAi00109167|WBPaper00046120  \n",
       "3285                     WBRNAi00075301|WBPaper00029258  \n",
       "3318                     WBRNAi00078335|WBPaper00027214  \n",
       "3365  WBRNAi00009253|WBPaper00005654 WBRNAi00026489|...  \n",
       "3410                     WBRNAi00075430|WBPaper00006132  \n",
       "3416                     WBRNAi00009310|WBPaper00005654  \n",
       "3528  WBRNAi00008327|WBPaper00005654 WBRNAi00024451|...  \n",
       "3594  WBRNAi00025512|WBPaper00006395 WBRNAi00026260|...  \n",
       "3646                     WBRNAi00078339|WBPaper00027214  \n",
       "3652  WBRNAi00024790|WBPaper00006395 WBRNAi00078676|...  \n",
       "3774                     WBRNAi00027137|WBPaper00006395  \n",
       "3778                     WBRNAi00008776|WBPaper00005654  \n",
       "3792  WBRNAi00024619|WBPaper00006395 WBRNAi00024620|...  \n",
       "3840                     WBRNAi00008581|WBPaper00005654  \n",
       "3856  WBRNAi00008892|WBPaper00005654 WBRNAi00025742|...  \n",
       "3950                     WBRNAi00025936|WBPaper00006395  \n",
       "4109  WBRNAi00025481|WBPaper00006395 WBRNAi00027336|...  \n",
       "4126  WBRNAi00025372|WBPaper00006395 WBRNAi00027316|...  \n",
       "4171  WBRNAi00025950|WBPaper00006395 WBRNAi00025951|...  \n",
       "4189  WBRNAi00008915|WBPaper00005654 WBRNAi00025787|...  \n",
       "4250                     WBRNAi00009393|WBPaper00005654  \n",
       "4271  WBRNAi00025094|WBPaper00006395 WBRNAi00027277|...  \n",
       "\n",
       "[100 rows x 5 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_location_variant.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## use reference list to extract only protein coding genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "cElegans_proteinCoding = cElegans_geneInfo[cElegans_geneInfo[\"type_of_gene\"] == \"protein-coding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_onlyProteinCoding = df.loc[df[\"WB gene ID\"].isin(cElegans_proteinCoding[\"dbXrefs\"].unique())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(53073, 5)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_onlyProteinCoding.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load the table for PMID # reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "paper_lookup = pd.read_csv('input/WB_paper_id_conversion_table', sep='~', header=None)\n",
    "paper_lookup.columns = ['info']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBPaper00000003 cgc3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBPaper00000005 cgc5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBPaper00000005 doi10.1163/187529275X00518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBPaper00000006 cgc6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBPaper00000007 cgc7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         info\n",
       "0                        WBPaper00000003 cgc3\n",
       "1                        WBPaper00000005 cgc5\n",
       "2  WBPaper00000005 doi10.1163/187529275X00518\n",
       "3                        WBPaper00000006 cgc6\n",
       "4                        WBPaper00000007 cgc7"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_lookup.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "## need to split the file and then create a dictionary for each paper id\n",
    "table = paper_lookup['info'].apply(lambda x: x.split(' '))\n",
    "table_dict = {}\n",
    "for i in range(len(table)):\n",
    "    table_dict[table[i][0]] = table[i][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'pmid14551910'"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_dict['WBPaper00006395']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create gmt by writing to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = 'wormbase_rnai_phenotype_greater4_%s.gmt'% str(datetime.date.today())[0:7].replace('-', '_')\n",
    "file = open(filename,'w+') \n",
    "terms = df[\"phenotype\"].unique()\n",
    "for term in terms:\n",
    "    df_byTerm = df_onlyProteinCoding.loc[df[\"phenotype\"] == term].copy()\n",
    "    \n",
    "    if df_byTerm.shape[0] > 4:\n",
    "        # convert gene name using the cElegans gene info table\n",
    "        change_name = lambda x: cElegans_geneInfo.loc[cElegans_geneInfo[\"dbXrefs\"] == x][\"Symbol\"].tolist()[0] if x in cElegans_geneInfo[\"dbXrefs\"].values else x\n",
    "        df_byTerm[\"Gene symbol\"] = df_byTerm[\"WB gene ID\"].apply(change_name)\n",
    "        # split splice variant names. some genes have format X.#.# where the second decimal describes a TX ID\n",
    "        split_splice = lambda x: (x.split('.')[0] + x.split('.')[1]) if len(x.split('.')) > 2 else x\n",
    "        df_byTerm[\"Gene symbol\"] = df_byTerm[\"Gene symbol\"].apply(split_splice)\n",
    "        \n",
    "        if len(df_byTerm[\"Gene symbol\"].unique()) > 4:\n",
    "            file.write(\"%s\" % term + \"_\" + df_byTerm.iloc[0, 3] + '\\t')\n",
    "            file.write(\"\\t\")\n",
    "            genes = df_byTerm[\"Gene symbol\"].unique()\n",
    "            for gene in genes:\n",
    "                file.write(\"%s\\t\" % gene)\n",
    "            file.write(\"\\n\")\n",
    "        \n",
    "file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get mean and median for each term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getTermStats(loaded_gmt, geneInfoTable, num_samplings, num_genes_in_sampling) :\n",
    "    # get the relevant terms and set a dictionary for them\n",
    "    term_set = loaded_gmt.iloc[:,0].apply(lambda x: x.split(\"\\t\")[0])\n",
    "    term_genes_dict = {term_set[i]: loaded_gmt.iloc[i,0].split(\"\\t\")[2:] for i in range(0, loaded_gmt.shape[0])}\n",
    "    term_rank_dict = {k: [] for k in term_set}\n",
    "    term_stats_dict = {k: [] for k in term_set}\n",
    "    \n",
    "    # get the total number of genes in the organism\n",
    "    n_genes = len(geneInfoTable[\"Symbol\"].unique())\n",
    "    \n",
    "    # number of genes per sampling, number of times to sample\n",
    "    genes_in_sampleing = num_genes_in_sampling\n",
    "    num_samplings = num_samplings\n",
    "    \n",
    "    for i,x in enumerate(range(0,num_samplings)):\n",
    "        \n",
    "        progressPercent = ((i+1)/len(range(0,num_samplings)))*100\n",
    "        \n",
    "        sys.stdout.write(\"Progress: %d%%  %d Out of %d   \\r\" % (progressPercent, (i+1), len(range(0,num_samplings))))\n",
    "        sys.stdout.flush()\n",
    "            \n",
    "        \n",
    "        randomSet = set(geneInfoTable[\"Symbol\"].sample(genes_in_sampleing).apply(lambda x: x.split('.')[0]))\n",
    "        \n",
    "        # initialize an array to hold the pvalue for each term\n",
    "        pvals = np.array([])\n",
    "        for key in term_rank_dict.keys():\n",
    "            \n",
    "            termSet = set(term_genes_dict[key])\n",
    "            # calculate p value using fisher exact test\n",
    "            # implemented using the formula found in the GeneOverap bioconductor package for R\n",
    "            pval_term = stat.fisher_exact([[n_genes - len(termSet.union(randomSet)), len(randomSet.difference(termSet))], [len(termSet.difference(randomSet)), len(termSet.intersection(randomSet))]])[1]\n",
    "            pvals = np.append(pvals, [pval_term])\n",
    "        \n",
    "        # sort the pvals and add the rank to the term_rank dict\n",
    "        sorted_pvals = pvals.argsort()\n",
    "        \n",
    "        # find the rank for each term by sorting the pvals array and getting the index of the key position in the\n",
    "        # sorted list. Need to +1 because the index begins with 0\n",
    "        \n",
    "        for j,key in enumerate(term_rank_dict.keys()):\n",
    "            term_rank_dict[key].append(np.where(sorted_pvals == j)[0][0] + 1)\n",
    "            \n",
    "    for key in term_rank_dict:\n",
    "        term_stats_dict[key] = [np.mean(term_rank_dict[key]), np.std(term_rank_dict[key])]\n",
    "            \n",
    "    \n",
    "    \n",
    "    return term_stats_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load gmt\n",
    "rnai_WB_gmt = pd.read_csv(\"wormbase_rnai_phenotype_greater4_2018_06.gmt\", sep=\"~\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>locomotion variant\\t\\this-34\\tleo-1\\tubc-12\\tn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>avoids bacterial lawn\\t\\this-34\\tpign-1\\tset-3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RAB-11 recycling endosome localization variant...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>body elongation defective\\t\\this-34\\this-11\\th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chromosome segregation variant\\t\\this-34\\tsmc-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   0\n",
       "0  locomotion variant\\t\\this-34\\tleo-1\\tubc-12\\tn...\n",
       "1  avoids bacterial lawn\\t\\this-34\\tpign-1\\tset-3...\n",
       "2  RAB-11 recycling endosome localization variant...\n",
       "3  body elongation defective\\t\\this-34\\this-11\\th...\n",
       "4  chromosome segregation variant\\t\\this-34\\tsmc-..."
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnai_WB_gmt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress: 100%  200 Out of 200   \r"
     ]
    }
   ],
   "source": [
    "iterations_200 = getTermStats(rnai_WB_gmt, cElegans_proteinCoding, num_samplings=200, num_genes_in_sampling=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ABa ABp EMS synchronous division early emb': [446.065, 174.11938655704023],\n",
       " 'ATP levels reduced': [515.425, 175.50610922415208],\n",
       " 'Aba ABp division axis defective early emb': [391.655, 200.22456386517615],\n",
       " 'Bacillus thuringiensis toxin hypersensitive': [189.135, 111.13917749830615],\n",
       " 'C lineage variant': [248.825, 72.95638680060848],\n",
       " 'DNA damage checkpoint variant': [554.725, 168.04359367437962],\n",
       " 'DNA repair variant': [282.915, 84.69443768630853],\n",
       " 'E lineage variant': [416.83, 173.67544760270522],\n",
       " 'EMS anterior extension fails early emb': [395.25, 181.13560527958052],\n",
       " 'Fluorouracil resistant': [479.1, 125.6230074468845],\n",
       " 'L1 arrest': [327.24, 222.3097217847209],\n",
       " 'L1 lethal': [284.46, 164.29253908805475],\n",
       " 'L2 arrest': [289.65, 112.80397820999045],\n",
       " 'L3 arrest': [598.365, 176.33273029985105],\n",
       " 'L4 lethal': [281.915, 89.70589598794496],\n",
       " 'M lineage variant': [501.695, 180.99768499900765],\n",
       " 'P granule defective': [239.99, 110.26082667928806],\n",
       " 'P granule localization defective': [243.255, 120.83848714296286],\n",
       " 'P0 spindle position defective early emb': [187.635, 81.07386616536799],\n",
       " 'P0 spindle rotation delayed early emb': [380.925, 154.99144936092443],\n",
       " 'P0 spindle rotation failure early emb': [385.62, 160.04738548317496],\n",
       " 'P1 aster defective early emb': [424.835, 176.77193718178233],\n",
       " 'P2 EMS synchronous division early emb': [297.12, 111.95648976276453],\n",
       " 'Q neuroblast lineage migration variant': [260.67, 120.2234631841888],\n",
       " 'RAB-11 recycling endosome localization variant': [175.325,\n",
       "  105.53008753431412],\n",
       " 'RAB-11 recycling endosome morphology variant': [179.075, 104.1842568481438],\n",
       " 'RNA degradation variant': [459.79, 142.5321223444035],\n",
       " 'RNA expression variant': [514.23, 167.021786303464],\n",
       " 'RNAi resistant': [187.95, 88.48908124734938],\n",
       " 'SDS sensitive dauer': [535.825, 168.4340653638687],\n",
       " 'UV induced apoptosis decreased': [525.995, 131.00089684807506],\n",
       " 'V lineages variant': [570.365, 184.09215565851795],\n",
       " 'VPC cell division variant': [406.905, 142.68481340002518],\n",
       " 'accumulated cell corpses': [258.125, 187.43086558782147],\n",
       " 'accumulated germline cell corpses': [169.935, 88.52237443155262],\n",
       " 'accumulated somatic cell corpses': [287.82, 100.1059318921711],\n",
       " 'actin cytoskeleton dynamics variant': [313.0, 109.29569982391804],\n",
       " 'actin cytoskeleton filament morphology variant': [252.505,\n",
       "  103.65452221200964],\n",
       " 'actin organization biogenesis variant': [407.545, 159.48409944254632],\n",
       " 'adherens junction variant': [543.695, 183.6931734578071],\n",
       " 'adult lethal': [193.67, 97.51159469519509],\n",
       " 'age associated fluorescence increased': [235.225, 114.7556725177453],\n",
       " 'aging variant': [365.235, 182.52953124083785],\n",
       " 'alae absent': [393.55, 185.5381295044229],\n",
       " 'alae morphology variant': [268.185, 150.28070659602318],\n",
       " 'alae secretion variant': [264.765, 119.27761640391712],\n",
       " 'alae variant': [532.845, 237.93879669990767],\n",
       " 'aldicarb hypersensitive': [519.965, 191.17772823998092],\n",
       " 'aldicarb resistant': [178.78, 94.23450323528002],\n",
       " 'alternative splicing variant': [251.66, 86.68249188850075],\n",
       " 'amphid phasmid morphology variant': [431.51, 210.21938516702022],\n",
       " 'anaphase bridging': [331.9, 258.33828597403055],\n",
       " 'anchor cell invasion variant': [437.975, 125.77783737606558],\n",
       " 'aneuploidy': [278.785, 72.8116664759158],\n",
       " 'annulae morphology variant': [428.91, 162.48880546056088],\n",
       " 'anoxia hypersensitive': [282.875, 119.52409537411275],\n",
       " 'anterior pharynx variant': [567.6, 126.44694539608301],\n",
       " 'antibody staining increased': [463.98, 139.84598528381142],\n",
       " 'antibody staining reduced': [397.13, 267.24687668895217],\n",
       " 'antibody staining variant': [526.98, 201.43917096731707],\n",
       " 'antibody subcellular localization variant': [223.165, 101.35885642113371],\n",
       " 'apoptosis fails to occur': [205.96, 96.86851087943903],\n",
       " 'apoptosis increased': [233.815, 165.20311974959793],\n",
       " 'apoptosis reduced': [197.31, 130.5095931339915],\n",
       " 'apoptosis variant': [179.24, 106.6058741345898],\n",
       " 'aster defective early emb': [304.035, 117.68710114111911],\n",
       " 'asymmetric cell division defective early emb': [336.715, 165.66545739833634],\n",
       " 'asymmetric cell division variant': [356.545, 148.59918564716295],\n",
       " 'autophagy increased': [210.81, 97.20166613798345],\n",
       " 'autophagy reduced': [304.735, 104.09771743414935],\n",
       " 'autophagy variant': [287.165, 104.63989571382417],\n",
       " 'avoids bacterial lawn': [179.51, 163.4417018389126],\n",
       " 'axon branching variant': [474.63, 141.28320176156822],\n",
       " 'axon fasciculation variant': [310.11, 221.22002147183693],\n",
       " 'axon guidance variant': [237.485, 105.70193836917089],\n",
       " 'axon midline crossing variant': [319.39, 213.96716547171437],\n",
       " 'axon outgrowth variant': [572.6, 140.6535459915604],\n",
       " 'bacterially unswollen': [606.665, 159.87242656255643],\n",
       " 'bag of worms': [532.08, 175.78123790666623],\n",
       " 'blistered': [394.47, 229.82149399044468],\n",
       " 'body axis development variant': [296.825, 104.37913764253851],\n",
       " 'body elongation defective': [275.525, 178.06178527410086],\n",
       " 'body elongation variant': [232.81, 80.29273877505985],\n",
       " 'body morphology variant': [289.085, 119.64120433613162],\n",
       " 'body muscle displaced': [380.985, 141.7478210590907],\n",
       " 'body muscle dystrophy': [240.11, 104.17906651530336],\n",
       " 'body vacuole': [354.335, 232.70544208290445],\n",
       " 'body wall muscle development variant': [413.76, 224.65474043518424],\n",
       " 'body wall muscle morphology variant': [387.86, 229.78076594876256],\n",
       " 'body wall muscle myosin organization defective': [190.335,\n",
       "  111.72592704918586],\n",
       " 'body wall muscle physiology variant': [514.935, 140.81910656938567],\n",
       " 'body wall muscle sarcomere morphology variant': [260.655,\n",
       "  206.12233254793136],\n",
       " 'branched adult alae': [403.185, 132.0017453483097],\n",
       " 'breaks in alae': [240.405, 117.47272438740833],\n",
       " 'brood size variant': [594.365, 179.5613593594123],\n",
       " 'cadmium hypersensitive': [199.38, 70.36153494630429],\n",
       " 'cadmium response variant': [444.34, 130.86838579274982],\n",
       " 'cell acidification defective': [422.085, 130.26756225169794],\n",
       " 'cell cell contacts variant four emb': [379.51, 247.8010893842075],\n",
       " 'cell corpse degradation variant': [307.76, 205.10485708534551],\n",
       " 'cell corpse number decreased': [441.06, 211.90098253665556],\n",
       " 'cell corpse number variant': [404.125, 126.60161679457336],\n",
       " 'cell cycle delayed': [308.17, 101.18365035913658],\n",
       " 'cell cycle progression variant': [257.725, 106.0212685030697],\n",
       " 'cell cycle slow early emb': [292.955, 184.41188403950545],\n",
       " 'cell cycle timing defective early emb': [284.73, 130.45185740341148],\n",
       " 'cell cycle variant': [271.845, 86.88475686217923],\n",
       " 'cell death variant': [239.325, 118.54370238439493],\n",
       " 'cell development variant': [400.985, 230.97466262557893],\n",
       " 'cell differentiation variant': [195.73, 83.3077853504701],\n",
       " 'cell division polarity variant': [398.685, 116.07396682719171],\n",
       " 'cell division precocious': [496.11, 223.24963135467883],\n",
       " 'cell division slow': [478.24, 185.80772965622285],\n",
       " 'cell division variant': [293.26, 103.5125712172198],\n",
       " 'cell fate specification variant': [238.65, 91.73656577395951],\n",
       " 'cell fate transformation': [346.71, 269.0364768948627],\n",
       " 'cell fusion variant': [322.725, 86.88089188653625],\n",
       " 'cell membrane morphology variant': [388.135, 151.38357498421024],\n",
       " 'cell membrane organization biogenesis variant': [202.79, 167.74941400791838],\n",
       " 'cell migration variant': [187.375, 79.23177629587765],\n",
       " 'cell morphology variant': [398.295, 195.10173749867016],\n",
       " 'cell polarity variant': [460.03, 226.27759301353726],\n",
       " 'cell position defective': [313.955, 98.33713934724764],\n",
       " 'cell proliferation increased': [195.195, 111.66748396467077],\n",
       " 'cell proliferation reduced': [318.455, 77.67089528903345],\n",
       " 'cell secretion variant': [319.9, 245.81234712682763],\n",
       " 'cell stress response variant': [331.105, 88.46510032210442],\n",
       " 'centriole biogenesis variant': [417.215, 112.36106431945187],\n",
       " 'centrosome attachment defective early emb': [305.705, 95.05565724879293],\n",
       " 'centrosome biogenesis variant': [578.495, 198.26928651457845],\n",
       " 'centrosome pair and associated pronuclear rotation variant': [411.555,\n",
       "  139.8087156617927],\n",
       " 'chemical hypersensitive': [363.615, 255.5846372045863],\n",
       " 'chemical resistant': [279.54, 122.66502517017635],\n",
       " 'chemical response variant': [451.84, 233.85083792879595],\n",
       " 'chromosome alignment variant': [562.11, 196.07431728811397],\n",
       " 'chromosome condensation variant': [235.61, 200.6429861719567],\n",
       " 'chromosome instability': [515.06, 172.28249011434679],\n",
       " 'chromosome morphology variant': [494.85, 225.24932297345535],\n",
       " 'chromosome segregation variant': [184.645, 106.51525231158212],\n",
       " 'chromosome segregation variant karyomeres early emb': [237.515,\n",
       "  93.00413848318794],\n",
       " 'clear': [234.005, 187.7885911736919],\n",
       " 'cleavage furrow defective early emb': [370.925, 190.52674188942612],\n",
       " 'cleavage furrow initiation defective early emb': [230.515,\n",
       "  96.95271927594398],\n",
       " 'cleavage furrow termination defective early emb': [346.54,\n",
       "  171.45217525595876],\n",
       " 'coelomocyte uptake defective': [248.495, 154.99429013676598],\n",
       " 'coenzyme Q depleted': [289.61, 120.944565400848],\n",
       " 'complex phenotype early emb': [342.885, 158.43141031689393],\n",
       " 'connected mitochondria': [561.335, 164.03991214030808],\n",
       " 'constipated': [210.925, 85.17047243616769],\n",
       " 'cord commissures fail to reach target': [202.54, 88.49914349868025],\n",
       " 'cortical dynamics defective early emb': [446.125, 247.17517952860882],\n",
       " 'cortical granule exocytosis variant': [263.345, 93.7280426286605],\n",
       " 'cuticle integrity variant': [504.13, 183.1840689033847],\n",
       " 'cuticle morphology variant': [261.895, 117.21413726594587],\n",
       " 'cytokinesis defective early emb': [443.44, 245.33633322441256],\n",
       " 'cytokinesis fails early emb': [223.155, 113.98180106929352],\n",
       " 'cytokinesis variant emb': [327.995, 217.0226139714477],\n",
       " 'cytoplasmic dynamics defective early emb': [539.555, 101.62498204181884],\n",
       " 'cytoplasmic processing body variant': [162.305, 80.79029629231471],\n",
       " 'cytoplasmic structure defective early emb': [433.255, 134.5144600963034],\n",
       " 'dauer arrest variant': [430.235, 155.73429222557246],\n",
       " 'dauer body morphology variant': [448.775, 156.69069651705553],\n",
       " 'dauer constitutive': [261.84, 165.47505673061423],\n",
       " 'dauer development variant': [258.5, 132.4795833326781],\n",
       " 'dauer formation variant': [484.34, 234.82213353940895],\n",
       " 'dauer induction variant': [300.625, 87.69956884158553],\n",
       " 'dauer lifespan extended': [115.82, 116.06557456886172],\n",
       " 'dead eggs laid': [278.04, 172.08767649079348],\n",
       " 'defecation cycle variant': [401.885, 145.71733519042954],\n",
       " 'dendrite development variant': [421.515, 229.5954698486013],\n",
       " 'development variant': [380.27, 209.02724009085514],\n",
       " 'developmental delay': [232.5, 86.39351827538916],\n",
       " 'developmental delay postembryonic': [406.185, 209.3276875499273],\n",
       " 'developmental timing variant': [386.955, 169.42692517719846],\n",
       " 'diakinesis progression during oogenesis variant': [258.855,\n",
       "  186.11357815860723],\n",
       " 'diakinesis region organization variant': [262.78, 125.64478341737869],\n",
       " 'diet induced life span variant': [539.965, 138.89414593495292],\n",
       " 'diplotene absent during oogenesis': [229.73, 157.80993980101508],\n",
       " 'diplotene progression during oogenesis variant': [215.12,\n",
       "  114.54652155347189],\n",
       " 'diplotene region organization variant': [204.65, 149.4716277425251],\n",
       " 'distal germline variant': [393.37, 130.03235405082845],\n",
       " 'distal tip cell development variant': [275.0, 124.200644120713],\n",
       " 'distal tip cell migration variant': [240.495, 173.33542619730108],\n",
       " 'division axis defective': [245.7, 90.0908985414176],\n",
       " 'drug resistant': [322.38, 90.60792239092561],\n",
       " 'dumpy': [178.95, 101.12278427733287],\n",
       " 'early embryonic lethal': [337.745, 223.07924595309174],\n",
       " 'early endosome localization variant': [283.065, 118.45104801140428],\n",
       " 'early endosome morphology variant': [283.05, 85.50489752055141],\n",
       " 'early larval arrest': [233.19, 176.0270544547059],\n",
       " 'early larval lethal': [265.025, 166.19396612091546],\n",
       " 'ectopic cell': [537.615, 156.4707856917706],\n",
       " 'ectopic cleavage furrows early emb': [376.54, 257.7077383393832],\n",
       " 'ectopic expression antibody': [446.03, 229.87239742953045],\n",
       " 'ectopic expression transgene': [283.205, 222.9404471490088],\n",
       " 'egg laying defective': [314.655, 237.8919838393047],\n",
       " 'egg laying variant': [174.725, 106.77780375621143],\n",
       " 'egg morphology variant': [613.935, 191.5841349772992],\n",
       " 'egg retention': [476.22, 99.16860188587918],\n",
       " 'egg size defective early emb': [303.215, 95.2665144476274],\n",
       " 'eggshell morphology variant': [549.37, 166.91115331217384],\n",
       " 'electrophysiology variant': [510.23, 166.9898113658435],\n",
       " 'embryo osmotic integrity defective early emb': [210.19, 110.59513506479387],\n",
       " 'embryo osmotic pressure sensitive early emb': [404.89, 236.4148428081452],\n",
       " 'embryonic arrest': [219.38, 154.50153915090942],\n",
       " 'embryonic cell organization biogenesis variant': [401.79,\n",
       "  145.16868085093284],\n",
       " 'embryonic development variant': [262.32, 126.44412837296953],\n",
       " 'embryonic lethal': [39.1, 54.056081988986215],\n",
       " 'embryonic lethal late emb': [318.96, 236.9720835879197],\n",
       " 'embryonic morphology variant': [218.31, 107.48145840097257],\n",
       " 'embryonic polarity variant': [452.135, 230.7117612411643],\n",
       " 'embryonic terminal arrest variable emb': [362.315, 159.97348459979236],\n",
       " 'endocytic transport defect': [464.31, 233.2381055916893],\n",
       " 'endocytic transport variant': [299.29, 117.78529577158602],\n",
       " 'endoderm absent': [397.065, 163.9793913118353],\n",
       " 'endomitotic oocytes': [349.87, 190.1912277156862],\n",
       " 'endoplasmic reticulum localization variant': [258.865, 65.78568822319943],\n",
       " 'endoplasmic reticulum morphology variant': [316.04, 95.51967546008518],\n",
       " 'endosome biogenesis variant': [609.2, 161.67526093994715],\n",
       " 'engulfment failure by killer cell': [431.19, 147.9063686931702],\n",
       " 'engulfment variant': [217.02, 110.24277572702894],\n",
       " 'enzyme activity reduced': [281.065, 115.80673890149916],\n",
       " 'epithelial development variant': [381.35, 150.1534132146186],\n",
       " 'epithelial morphology variant': [397.645, 220.71118905710242],\n",
       " 'everted vulva': [251.74, 119.985384109899],\n",
       " 'exaggerated asynchrony early emb': [235.76, 103.32880721270327],\n",
       " 'excess coelomocytes': [587.415, 154.44041820391448],\n",
       " 'excess hypodermal cells': [504.56, 180.92674318629625],\n",
       " 'excess intestinal cells': [212.905, 163.72197767862446],\n",
       " 'excess maternal pronucleus early emb': [404.54, 229.2604379303154],\n",
       " 'excess pharyngeal cells': [554.135, 176.80643872608258],\n",
       " 'excess seam cells': [549.185, 228.1446487976433],\n",
       " 'excessive blebbing early emb': [281.35, 172.60106459694853],\n",
       " 'excretory canal morphology variant': [423.88, 185.3180120765383],\n",
       " 'exploded through vulva': [190.075, 143.9744052774659],\n",
       " 'expulsion variant': [510.785, 152.18609915166365],\n",
       " 'extended life span': [159.35, 127.26463570057473],\n",
       " 'extra cell division postembryonic': [440.3, 165.61195609013257],\n",
       " 'extra distal tip cells': [318.83, 94.91175427732858],\n",
       " 'extra excretory cells': [459.33, 154.71377152664851],\n",
       " 'extra gonad arms': [388.58, 116.2624771798709],\n",
       " 'extra linker cells': [581.09, 129.68385365958247],\n",
       " 'failure to hatch': [240.395, 103.24412319836901],\n",
       " 'fat associated bodies increased': [490.885, 197.9816955554225],\n",
       " 'fat associated body size decreased': [557.335, 161.61142526133477],\n",
       " 'fat associated body size increased': [320.985, 87.32052894365677],\n",
       " 'fat content increased': [178.235, 102.02803426019733],\n",
       " 'fat content reduced': [167.44, 175.52597072798088],\n",
       " 'feeding reduced': [281.35, 75.05469672179083],\n",
       " 'feminization of XX and XO animals': [417.6, 158.4692399174048],\n",
       " 'feminization of germline': [510.475, 183.89793194867636],\n",
       " 'fertility reduced': [452.705, 230.52424595907476],\n",
       " 'fertility variant': [232.515, 69.89227264154457],\n",
       " 'fewer coelomocytes': [318.82, 89.4188324683341],\n",
       " 'fewer eggs laid': [507.625, 241.402991644677],\n",
       " 'fewer germ cells': [263.43, 136.2293841283884],\n",
       " 'fewer seam cells': [248.07, 109.86507679877168],\n",
       " 'foraging behavior variant': [212.195, 141.69751929726928],\n",
       " 'frequency body bend reduced': [476.39, 187.94240580560844],\n",
       " 'gametogenesis variant': [472.5, 200.19185797629234],\n",
       " 'gamma ray induced apoptosis decreased': [492.625, 118.02870148823972],\n",
       " 'gamma ray induced apoptosis increased': [335.69, 174.86819007469597],\n",
       " 'gastrulation variant': [390.295, 225.80597417916115],\n",
       " 'gene expression level high': [442.285, 193.11321491549978],\n",
       " 'gene expression level reduced': [224.765, 117.29756934821795],\n",
       " 'gene expression variant': [286.99, 117.40655816435469],\n",
       " 'general pace of development defective early emb': [302.085,\n",
       "  142.36901971636948],\n",
       " 'germ cell compartment anucleate': [218.92, 146.33305709920776],\n",
       " 'germ cell compartment expansion absent': [272.875, 179.13709659085134],\n",
       " 'germ cell compartment expansion variant': [192.335, 153.04928217734312],\n",
       " 'germ cell compartment large': [242.34, 106.33910099300257],\n",
       " 'germ cell compartment morphology variant': [181.84, 138.42674741537488],\n",
       " 'germ cell compartment multinucleate': [231.755, 167.32753800555363],\n",
       " 'germ cell compartment size variant': [202.3, 118.27146739598693],\n",
       " 'germ cell compartment small': [279.77, 178.62918882422323],\n",
       " 'germ cell development variant': [292.06, 230.92233846035768],\n",
       " 'germ cell differentiation variant': [510.99, 153.09663582195398],\n",
       " 'germ cell hypersensitive ionizing radiation': [283.08, 117.43501862732427],\n",
       " 'germ cell ionizing radiation response variant': [328.195, 163.7202399674518],\n",
       " 'germ cell mitosis variant': [466.78, 233.7955337469046],\n",
       " 'germ cell morphology variant': [204.03, 80.02667742696806],\n",
       " 'germ cell partition absent': [540.445, 185.55909294615554],\n",
       " 'germ cell partition morphology variant': [247.69, 150.41198057335725],\n",
       " 'germ cell resistant ionizing radiation': [253.5, 109.38157980208551],\n",
       " 'germline nuclear positioning variant': [231.72, 210.21670152487886],\n",
       " 'germline proliferation variant': [229.615, 127.52433797122806],\n",
       " 'germline transgene silencing variant': [491.82, 251.50083021731757],\n",
       " 'giant oocytes': [204.455, 81.58613837533922],\n",
       " 'golgi apparatus localization variant': [461.405, 107.27940610853511],\n",
       " 'golgi apparatus morphology variant': [328.535, 76.420669815175],\n",
       " 'gonad arm morphology variant': [208.46, 100.63790737093056],\n",
       " 'gonad degenerate': [481.965, 211.08904229021456],\n",
       " 'gonad development variant': [247.945, 208.92925112343653],\n",
       " 'gonad displaced': [261.415, 90.8313975175985],\n",
       " 'gonad migration variant': [231.24, 156.59078006064087],\n",
       " 'gonad morphology variant': [230.22, 166.20912008671485],\n",
       " 'gonad sheath contraction rate reduced': [326.37, 79.90127095359624],\n",
       " 'gonad small': [298.04, 194.32817191544822],\n",
       " 'gonad vesiculated': [235.085, 192.26585181721686],\n",
       " 'growth rate variant': [198.335, 95.48325913478237],\n",
       " 'growth variant': [256.465, 229.21941186339345],\n",
       " 'gut granule biogenesis reduced': [520.145, 192.47276164434282],\n",
       " 'gut granule development variant': [384.675, 151.08715820677813],\n",
       " 'gut granules enlarged': [481.085, 117.49824583796985],\n",
       " 'halothane hypersensitive': [437.34, 84.32238374239667],\n",
       " 'hatching variant': [521.73, 168.6756861554148],\n",
       " 'head morphology variant': [460.23, 151.3428131759153],\n",
       " 'head notched': [326.395, 80.84997820036811],\n",
       " 'high incidence male progeny': [271.155, 178.751282442952],\n",
       " 'high temperature induced L1 arrest': [572.68, 196.0282826532947],\n",
       " 'histone methylation variant': [253.94, 108.54826760478493],\n",
       " 'hyperactive': [224.505, 93.16039917797691],\n",
       " 'hyperinduction of antimicrobial peptide expression after infection': [279.89,\n",
       "  114.1819946401358],\n",
       " 'hypersensitive to protein aggregation induced paralysis': [412.37,\n",
       "  224.00297564987838],\n",
       " 'hypodermis disorganized': [438.78, 126.47870018307431],\n",
       " 'hypoosmotic shock hypersensitive': [442.485, 221.98689099809474],\n",
       " 'hypoxia hypersensitive': [580.25, 183.45173070865263],\n",
       " 'hypoxia resistant': [428.045, 239.1396097993806],\n",
       " 'increased levels of reactive oxygen species': [442.78, 96.24443672233737],\n",
       " 'increased pathogen accumulation': [253.95, 100.03558116990175],\n",
       " 'inhibition of oocyte maturation variant': [217.6, 95.24405493257834],\n",
       " 'integrity of membranous organelles defective early emb': [226.09,\n",
       "  93.77138102854197],\n",
       " 'interphase entry defective early emb': [607.6, 155.48803812512395],\n",
       " 'intestinal cell morphology variant': [254.365, 83.83520605926844],\n",
       " 'intestinal degeneration': [451.27, 222.7188072435734],\n",
       " 'intestinal development variant': [186.335, 83.25955065336348],\n",
       " 'intestinal morphology variant': [398.92, 231.98183894434496],\n",
       " 'intestinal physiology variant': [322.655, 175.06280580123237],\n",
       " 'intestinal vacuole': [363.63, 248.59948330597953],\n",
       " 'isoamyl alcohol chemotaxis defective': [584.415, 192.61524543763403],\n",
       " 'kinetochore organization variant': [259.0, 117.77958227129183],\n",
       " 'kinker': [249.65, 72.72047510845897],\n",
       " 'large cytoplasmic granules early emb': [412.5, 261.46013080391435],\n",
       " 'larval arrest': [94.455, 84.13464194373208],\n",
       " 'larval development retarded': [312.26, 103.63195646131554],\n",
       " 'larval lethal': [140.27, 79.63106868553253],\n",
       " 'late embryonic arrest': [225.205, 112.4584944546209],\n",
       " 'late endosome localization variant': [240.97, 103.47815759859664],\n",
       " 'late endosome morphology variant': [332.335, 70.60653493126539],\n",
       " 'late larval arrest': [192.315, 65.68695285214561],\n",
       " 'late larval lethal': [351.625, 211.330202231011],\n",
       " 'lawn leaving behavior variant': [418.3, 116.19431999887087],\n",
       " 'lethal': [207.755, 170.56284758117752],\n",
       " 'levamisole resistant': [426.815, 202.22547014409437],\n",
       " 'level of protein expression variant': [355.52, 175.7054057221917],\n",
       " 'level of transgene expression variant': [233.22, 116.29966294018223],\n",
       " 'life span variant': [263.09, 215.5868546549163],\n",
       " 'lima bean arrest': [253.435, 129.8881664163445],\n",
       " 'linker cell migration variant': [201.015, 86.16220038392706],\n",
       " 'lipid composition variant': [355.41, 177.29264479949526],\n",
       " 'lipid depleted': [340.495, 156.16715395690605],\n",
       " 'lipid metabolism variant': [380.745, 205.40021902373914],\n",
       " 'locomotion reduced': [232.84, 124.05282906890919],\n",
       " 'locomotion variant': [130.77, 133.11024415874235],\n",
       " 'locomotor coordination variant': [450.475, 126.37048458797648],\n",
       " 'long': [324.635, 234.63991939778705],\n",
       " 'long defecation cycle': [507.085, 215.59674806220988],\n",
       " 'loose mitotic furrow early emb': [379.245, 123.26355899048185],\n",
       " 'loss of asymmetry AWC': [249.82, 95.19883192560715],\n",
       " 'lumpy': [587.205, 149.07519235271843],\n",
       " 'lysosome morphology variant': [377.53, 133.99589210121331],\n",
       " 'lysosome organization biogenesis variant': [514.45, 149.2340024927295],\n",
       " 'lysosome-related organelle localization variant': [493.735,\n",
       "  191.62266247758902],\n",
       " 'lysosome-related organelle morphology variant': [178.135,\n",
       "  109.21623860488879],\n",
       " 'mRNA export variant': [436.22, 101.10989862520879],\n",
       " 'mRNA expression ectopic': [550.07, 139.92753517446093],\n",
       " 'mRNA expression variant': [294.095, 111.1719208028718],\n",
       " 'mRNA levels increased': [307.645, 219.9889746669137],\n",
       " 'mRNA levels reduced': [224.29, 144.50718286645824],\n",
       " 'mRNA surveillance defective': [206.315, 112.90547274158149],\n",
       " 'male gonad development variant': [204.16, 114.62239920713577],\n",
       " 'male ray morphology variant': [225.625, 76.87687802584077],\n",
       " 'male response to contact defective': [237.495, 86.63803999975993],\n",
       " 'male somatic gonad development variant': [226.425, 153.5140526955106],\n",
       " 'male sterile': [281.4, 95.1797772638705],\n",
       " 'male tail development variant': [279.575, 111.99095666615229],\n",
       " 'male tail morphology variant': [292.705, 124.41220991124624],\n",
       " 'male tail sensory ray development variant': [423.045, 118.54869453098165],\n",
       " 'masculinization of germline': [273.64, 159.88230170972645],\n",
       " 'maternal effect lethal emb': [398.31, 193.98026677989697],\n",
       " 'maternal effect sterile': [408.725, 112.66707316248169],\n",
       " 'maternal pronucleus indistinct early emb': [560.565, 164.1805889105043],\n",
       " 'maternal sterile': [111.38, 94.74500303446088],\n",
       " 'mechanosensation variant': [495.175, 113.11699419185429],\n",
       " 'meiosis defective early emb': [383.77, 169.1063189239243],\n",
       " 'meiosis metaphase to anaphase transition block': [366.205,\n",
       "  177.84013319551917],\n",
       " 'meiosis variant': [195.725, 111.54160378531411],\n",
       " 'meiotic chromosome segregation variant': [470.425, 259.77171588723814],\n",
       " 'meiotic progression during oogenesis variant': [260.28, 122.55689127911167],\n",
       " 'meiotic progression prophase variant': [392.95, 133.73517674867745],\n",
       " 'meiotic sister chromatid segregation variant': [518.935, 155.94784633011128],\n",
       " 'meiotic spindle defective': [312.07, 102.70051168324333],\n",
       " 'meiotic spindle defective early emb': [579.65, 146.55305353352418],\n",
       " 'meiotic spindle positioning defective': [377.495, 148.6553731790412],\n",
       " 'membrane trafficking variant': [346.615, 172.93515193563164],\n",
       " 'miRNA expression variant': [403.92, 186.82519530298907],\n",
       " 'microtubule nucleation variant': [604.3, 165.87239071044945],\n",
       " 'microtubule organization biogenesis variant': [481.635, 255.358614060697],\n",
       " 'microtubule polymerization variant': [392.375, 119.0744908660121],\n",
       " 'mid larval arrest': [312.02, 89.66230869211432],\n",
       " 'missing distal tip cells': [313.24, 162.5588890217942],\n",
       " 'mitochondria alignment variant': [191.695, 104.06590207652073],\n",
       " 'mitochondria morphology variant': [229.52, 173.62905747598816],\n",
       " 'mitochondrial DNA copy number variant': [434.72, 127.27851979026154],\n",
       " 'mitochondrial DNA repair variant': [285.585, 93.79500399808084],\n",
       " 'mitochondrial content increased': [624.15, 132.55458309692654],\n",
       " 'mitochondrial membrane potential reduced': [414.515, 173.82033763343114],\n",
       " 'mitochondrial metabolism variant': [265.355, 113.92615579839425],\n",
       " 'mitosis variant': [371.0, 265.5212797498536],\n",
       " 'mitotic chromosome segregation variant': [215.42, 85.80410013513341],\n",
       " 'mitotic spindle defective early emb': [216.94, 112.51620505509418],\n",
       " 'molt defect': [176.795, 94.41182645728236],\n",
       " 'morphology variant': [265.275, 120.16330294644868],\n",
       " 'motor neuron morphology variant': [405.85, 168.89815718355248],\n",
       " 'multiple alae': [504.085, 217.1778482603601],\n",
       " 'multiple anchor cells': [264.535, 100.15886768030079],\n",
       " 'multiple cytoplasmic cavities early emb': [227.655, 100.88843330630128],\n",
       " 'multiple nuclei early emb': [206.575, 106.91952288988199],\n",
       " 'multiple nuclei oocyte': [198.005, 82.06579662076035],\n",
       " 'multivulva': [257.67, 198.4660452067305],\n",
       " 'muscle arm development defective': [404.535, 113.36784718340557],\n",
       " 'muscle attachment variant': [405.165, 128.4184090191122],\n",
       " 'muscle cell attachment variant': [478.135, 233.08574983254553],\n",
       " 'muscle cell morphology variant': [513.825, 232.20397148843082],\n",
       " 'muscle detached': [253.57, 120.65796741201967],\n",
       " 'muscle excess': [461.29, 201.64584275407216],\n",
       " 'muscle missing': [259.185, 108.96334601598832],\n",
       " 'muscle system morphology variant': [561.01, 164.08534334302988],\n",
       " 'myopodia present': [422.795, 179.19822257768072],\n",
       " 'necrotic cell death variant': [291.625, 123.5496433624962],\n",
       " 'neuron cell body absent misplaced': [327.54, 85.54500803670545],\n",
       " 'neuron degeneration': [383.085, 150.62532248928133],\n",
       " 'neuron degeneration variant': [315.38, 109.08590926421249],\n",
       " 'neuron development variant': [558.0, 204.72911370882258],\n",
       " 'neuron number variant': [313.125, 98.70784859878164],\n",
       " 'neuron positioning variant': [301.67, 139.01133442996655],\n",
       " 'nicotine hypersensitive': [227.955, 107.53024214145526],\n",
       " 'nicotine resistant': [582.19, 172.14012867428676],\n",
       " 'no Intestine': [204.3, 98.10239548553338],\n",
       " 'no gut granules': [429.025, 181.66596922648998],\n",
       " 'no induction of antimicrobial peptide expression after infection': [205.24,\n",
       "  173.78237079750062],\n",
       " 'no oocytes': [189.385, 107.6179203246374],\n",
       " 'no pharynx': [383.69, 206.40061506691302],\n",
       " 'no polar body formation': [380.8, 133.11866886353695],\n",
       " 'nonsense mRNA accumulation': [302.405, 214.00147423557624],\n",
       " 'nuclear anchoring variant': [418.98, 126.91406383848876],\n",
       " 'nuclear appearance variant': [244.915, 218.29236765173445],\n",
       " 'nuclear envelope assembly variant early emb': [340.23, 197.0234937767575],\n",
       " 'nuclear envelope breakdown variant': [407.565, 159.83230516700934],\n",
       " 'nuclear fallout': [323.825, 211.64038928096878],\n",
       " 'nuclear morphology variation early emb': [316.7, 169.8620616853569],\n",
       " 'nuclear number variant': [236.535, 108.24245366306143],\n",
       " 'nuclear pore variant': [426.135, 116.84445547393338],\n",
       " 'nuclear positioning variant': [268.335, 124.08510295357779],\n",
       " 'nuclei enlarged': [197.55, 108.56084699374817],\n",
       " 'nuclei small': [250.09, 193.55767073407344],\n",
       " 'nucleolus variant': [585.115, 193.09485175685032],\n",
       " 'nucleus defective early emb': [385.085, 125.18125967971403],\n",
       " 'nucleus reforms cell division remnant early emb': [264.665,\n",
       "  118.7415798067383],\n",
       " 'octanol chemotaxis defective': [435.16, 136.63419923284215],\n",
       " 'one cell arrest early emb': [211.18, 109.24608734412413],\n",
       " 'one cell shape defective early emb': [525.845, 165.44497869382437],\n",
       " 'oocyte accumulation': [452.95, 226.08924675888503],\n",
       " 'oocyte meiotic maturation variant': [532.15, 170.55596588803337],\n",
       " 'oocyte morphology variant': [238.935, 197.81802439363307],\n",
       " 'oocyte number decreased': [198.1, 143.9865618729748],\n",
       " 'oocyte physiology variant': [532.605, 177.16223913407734],\n",
       " 'oocyte septum formation variant': [216.895, 143.32869208570906],\n",
       " 'oocytes disorganized': [187.33, 82.97488234399611],\n",
       " 'oocytes lack nucleus': [194.705, 104.12424297443896],\n",
       " 'oocytes small': [264.035, 184.36890674677224],\n",
       " 'oocytes stacked': [523.78, 171.05297892758256],\n",
       " 'oogenesis variant': [256.795, 204.09042842573484],\n",
       " 'organism UV hypersensitive': [387.92, 147.97517899972277],\n",
       " 'organism development variant': [140.505, 100.5550096961857],\n",
       " 'organism electrophilic stress hypersensitive': [251.93, 117.31860508887753],\n",
       " 'organism heat hypersensitive': [370.965, 186.73447398646024],\n",
       " 'organism heat response variant': [539.49, 226.4073097318194],\n",
       " 'organism homeostasis metabolism variant': [285.755, 87.23889599828736],\n",
       " 'organism hypersensitive ionizing radiation': [225.975, 100.24786469047608],\n",
       " 'organism hypertonic lethality increased': [294.57, 153.2395024137053],\n",
       " 'organism metabolism processing variant': [319.37, 97.20042746819584],\n",
       " 'organism morphology variable': [509.47, 164.69070738812192],\n",
       " 'organism morphology variant': [175.82, 128.3596026793477],\n",
       " 'organism osmotic stress response variant': [446.7, 238.15190530415666],\n",
       " 'organism oxidative stress resistant': [276.405, 207.47867595249397],\n",
       " 'organism oxidative stress response hypersensitive': [204.815,\n",
       "  71.04682100558757],\n",
       " 'organism pathogen response variant': [276.515, 112.95844269022126],\n",
       " 'organism starvation response variant': [310.095, 101.41639894514103],\n",
       " 'organism stress response variant': [488.19, 195.78335960954396],\n",
       " 'organism temperature response variant': [300.79, 109.92149880710323],\n",
       " 'organism toxic chemical response variant': [506.17, 197.40912111652796],\n",
       " 'ovulation variant': [469.31, 237.9892726574036],\n",
       " 'oxygen consumption decreased': [420.805, 159.07324405757242],\n",
       " 'oxygen response variant': [457.785, 159.25516875442378],\n",
       " 'pace of P lineage defective early emb': [272.34, 113.17629787194844],\n",
       " 'pachytene arrested germline nuclei': [449.2, 112.37842319591425],\n",
       " 'pachytene progression during oogenesis variant': [324.275,\n",
       "  254.82232903534964],\n",
       " 'pachytene region organization variant': [190.0, 126.55848450420066],\n",
       " 'pale': [261.365, 117.4620014089663],\n",
       " 'paralyzed': [259.25, 182.32473090615],\n",
       " 'paralyzed arrested elongation two fold': [288.675, 123.64642888090218],\n",
       " 'paraquat hypersensitive': [229.54, 122.48754385650811],\n",
       " 'paraquat resistant': [235.905, 92.97212472026226],\n",
       " 'passage through meiosis defective early emb': [206.92, 100.71118905067104],\n",
       " 'patchy coloration': [207.95, 106.77432978014893],\n",
       " 'pathogen adherence increased': [406.125, 140.34289926818528],\n",
       " 'pathogen induced death increased': [332.05, 162.2726948688534],\n",
       " 'pathogen induced gene expression variant': [253.685, 112.85635017578763],\n",
       " 'pathogen infected pharynx increase': [244.88, 69.773172494878],\n",
       " 'pathogen load variant': [398.195, 123.72056003348837],\n",
       " 'pathogen resistance increased': [285.99, 201.6512581165811],\n",
       " 'pathogen susceptibility increased': [190.465, 109.27473987614887],\n",
       " 'pattern of transgene expression variant': [119.295, 86.06333699665613],\n",
       " 'pattern protein expression variant': [395.525, 183.22829851035564],\n",
       " 'peptide uptake by intestinal cell decreased': [214.44, 106.93496341234705],\n",
       " 'persistence of paternal mitochondria': [373.385, 155.95780446967058],\n",
       " 'phagosome maturation defective': [463.945, 215.13751875254115],\n",
       " 'pharyngeal development variant': [283.93, 183.34940714384652],\n",
       " 'pharyngeal morphology variant': [233.53, 106.9274478326309],\n",
       " 'pharyngeal pumping rate variant': [592.48, 170.51944053391685],\n",
       " 'pharyngeal pumping reduced': [324.8, 169.1507611570223],\n",
       " 'pharyngeal pumping variant': [314.275, 95.21454392580999],\n",
       " 'pharynx unattached': [490.015, 233.21958488729032],\n",
       " 'phasmid socket absent': [280.74, 114.1829339262221],\n",
       " 'pleiotropic defects severe early emb': [203.845, 107.31030227801988],\n",
       " 'polar body defective early emb': [230.42, 97.9162070343822],\n",
       " 'polar body extrusion defective early emb': [373.065, 174.3270511854084],\n",
       " 'polar body number defective early emb': [274.3, 101.82853234727484],\n",
       " 'polar body reabsorbed one two early emb': [379.825, 177.31332261000583],\n",
       " 'polar body size defective early emb': [284.095, 85.8436717236629],\n",
       " 'population fitness variant': [316.02, 174.04605597369908],\n",
       " 'pore forming toxin hypersensitive': [218.55, 163.64164353855654],\n",
       " 'posterior body morphology variant': [529.02, 205.06984078601124],\n",
       " 'posterior body wall contraction defecation cycle variant': [295.49,\n",
       "  88.44382341350921],\n",
       " 'posterior reversed vulval lineage': [400.81, 139.12995328109616],\n",
       " 'precocious alae': [492.78, 210.65567070458843],\n",
       " 'precocious heterochronic variations': [603.43, 147.7016083189347],\n",
       " 'precocious seam cell fusion': [527.3, 231.18483946833538],\n",
       " 'programmed cell death variant': [285.695, 108.9687201677619],\n",
       " 'progressive paralysis': [441.065, 141.8453057912034],\n",
       " 'pronuclear breakdown asynchronous early emb': [306.85, 115.13960005141585],\n",
       " 'pronuclear envelope defective early emb': [527.815, 181.43632705442425],\n",
       " 'pronuclear migration defective early emb': [203.7, 102.13799488926735],\n",
       " 'pronuclear migration failure early emb': [518.84, 207.13515973875607],\n",
       " 'pronuclear nuclear appearance defective early emb': [378.15,\n",
       "  202.16349200585154],\n",
       " 'pronuclear size defective early emb': [219.8, 105.84918516455382],\n",
       " 'pronuclei meet centrally early emb': [439.68, 235.16278957352074],\n",
       " 'pronucleus centrosomes defective early emb': [251.41, 113.9890867583384],\n",
       " 'pronucleus formation failure early emb': [415.44, 130.40673448867585],\n",
       " 'proteasome core activity variant': [229.9, 85.84451059910587],\n",
       " 'protein DNA interaction variant': [309.905, 100.03687307688101],\n",
       " 'protein acetylation variant': [440.385, 127.88196422873712],\n",
       " 'protein aggregation variant': [162.555, 140.9466458451566],\n",
       " 'protein complex organization variant': [232.12, 67.75895217607781],\n",
       " 'protein degradation variant': [183.635, 101.44590565912455],\n",
       " 'protein expression absent': [520.825, 206.39843598002383],\n",
       " 'protein expression ectopic': [249.135, 112.46100112928036],\n",
       " 'protein expression increased': [351.075, 169.4395448972878],\n",
       " 'protein expression reduced': [229.97, 165.68723276100664],\n",
       " 'protein expression variant': [243.285, 115.70654162578708],\n",
       " 'protein metabolism variant': [322.4, 164.99324228585849],\n",
       " 'protein methylation variant': [436.875, 137.88233887993053],\n",
       " 'protein phosphorylation increased': [259.385, 114.56219609888771],\n",
       " 'protein phosphorylation reduced': [221.62, 100.82646279623222],\n",
       " 'protein protein interaction variant': [568.925, 149.830168440805],\n",
       " 'protein secretion variant': [306.045, 99.37823189713129],\n",
       " 'protein subcellular localization variant': [226.495, 151.00235089229574],\n",
       " 'protein transport variant': [374.295, 155.53201591633797],\n",
       " 'protein ubiquitination variant': [443.605, 182.78372185454592],\n",
       " 'protruding vulva': [188.925, 179.02541544428826],\n",
       " 'proximal germ cell proliferation variant': [514.36, 156.70411736773224],\n",
       " 'pseudocleavage absent early emb': [457.155, 153.28193296993615],\n",
       " 'pseudocleavage defective early emb': [544.635, 161.40595334435469],\n",
       " 'pseudocleavage exaggerated early emb': [267.64, 96.02952879192942],\n",
       " 'pseudocoelom morphology variant': [553.175, 225.1415207708254],\n",
       " 'quiescence variant': [551.52, 113.18184306680996],\n",
       " 'rachis absent': [229.385, 114.23789552946081],\n",
       " 'rachis morphology variant': [218.885, 195.0505877330289],\n",
       " 'rachis narrow': [242.02, 217.3840601332122],\n",
       " 'rachis wide': [218.105, 100.21219474195742],\n",
       " 'radiation induced germ cell apoptosis resistant': [416.44,\n",
       "  123.57595397163641],\n",
       " 'ray ectopic': [292.895, 77.22541016401273],\n",
       " 'rays displaced': [378.335, 175.2349644762711],\n",
       " 'rays fused': [440.465, 172.74069808530936],\n",
       " 'rays missing': [285.97, 104.50286646786297],\n",
       " 'receptor mediated endocytosis defective': [134.0, 108.68403746640993],\n",
       " 'reduced brood size': [150.565, 147.07462655060525],\n",
       " 'reduced levels of reduced glutathione': [285.78, 87.5289757737402],\n",
       " 'reproductive system development variant': [405.545, 117.4914378795323],\n",
       " 'reproductive system morphology variant': [520.97, 145.49631988473112],\n",
       " 'reproductive system physiology variant': [326.32, 100.32446162327511],\n",
       " 'resistant to protein aggregation induced paralysis': [244.235,\n",
       "  103.94084748067047],\n",
       " 'retarded heterochronic variations': [518.3, 165.2006053257675],\n",
       " 'roller': [230.025, 100.79010058036455],\n",
       " 'rudimentary gonad': [299.45, 99.37070745446064],\n",
       " 'scrawny': [447.43, 123.18877830386988],\n",
       " 'seam cell fusion variant': [216.425, 74.70317513332348],\n",
       " 'seam cell morphology variant': [274.47, 97.60276174371296],\n",
       " 'short': [357.805, 144.3743986134661],\n",
       " 'short defecation cycle': [412.365, 234.23618801329567],\n",
       " 'shortened life span': [190.12, 108.33926158138608],\n",
       " 'sick': [144.025, 97.37748392210594],\n",
       " 'sister cell division timing asynchrony reduced': [210.83,\n",
       "  141.56366447644677],\n",
       " 'sister chromatid segregation defective early emb': [265.315,\n",
       "  192.4332241973823],\n",
       " 'slow development': [501.61, 215.91062479646527],\n",
       " 'slow embryonic development': [204.37, 109.59203027592837],\n",
       " 'slow growth': [63.86, 80.59727290671813],\n",
       " 'sluggish': [276.005, 229.76258393176207],\n",
       " 'small': [188.57, 148.56249560370208],\n",
       " 'somatic gonad development variant': [354.85, 212.26016465648942],\n",
       " 'somatic gonad morphology variant': [391.335, 115.88926945580424],\n",
       " 'somatic gonad precursor symmetrical sisters': [290.695, 76.85025683106076],\n",
       " 'somatic transgene silencing variant': [179.765, 80.39421481051978],\n",
       " 'sperm absent': [336.1, 157.13341465137196],\n",
       " 'sperm excess': [546.17, 148.68332488883885],\n",
       " 'sperm morphology variant': [279.135, 97.32690673703752],\n",
       " 'spermatheca physiology variant': [468.115, 143.92884274876943],\n",
       " 'spermatogenesis variant': [253.5, 112.62726135354619],\n",
       " 'spindle assembly defective early emb': [357.18, 162.5809878183793],\n",
       " 'spindle defective early emb': [368.405, 163.4963332157636],\n",
       " 'spindle elongation integrity defective early emb': [426.51,\n",
       "  217.95141637530142],\n",
       " 'spindle orientation defective early emb': [256.305, 122.80094451998323],\n",
       " 'spindle orientation variant': [509.74, 211.11656116941654],\n",
       " 'spindle orientation variant AB or P1 early emb': [286.29,\n",
       "  172.45627242869423],\n",
       " 'spindle oscillation variant': [429.115, 140.3578703707063],\n",
       " 'spindle position defective early emb': [265.78, 117.80658555445872],\n",
       " 'spindle position orientation defective early emb': [252.085,\n",
       "  132.55752628576016],\n",
       " 'spindle rotation defective early emb': [290.075, 116.42456516989874],\n",
       " 'spontaneous mutation rate increased': [279.81, 197.32669332860164],\n",
       " 'spontaneous spicule protraction': [582.86, 155.74694346920583],\n",
       " 'sterile': [151.155, 132.5130596394182],\n",
       " 'sterile F1': [205.755, 84.58182414088739],\n",
       " 'sterile progeny': [112.545, 95.1393608082375],\n",
       " 'stuffed pharynx': [562.595, 143.8673728647326],\n",
       " 'supernumerary centrosomes': [479.585, 120.93300945151411],\n",
       " 'swollen intestine': [438.325, 109.33777652302977],\n",
       " 'synapse density variant': [298.575, 113.11226447649256],\n",
       " 'synapsis defective': [494.08, 181.13070860569172],\n",
       " 'synchronous second division early emb': [377.495, 173.4810075339661],\n",
       " 'tail morphology variant': [254.62, 105.00916912346274],\n",
       " 'tail swelling': [424.705, 105.28004547396434],\n",
       " 'terminal arrest variable': [285.345, 110.64034515040163],\n",
       " 'thermotolerance increased': [251.455, 95.15470548007598],\n",
       " 'thermotolerance reduced': [338.425, 156.26936480001447],\n",
       " 'thin': [210.8, 156.24938399878573],\n",
       " 'thrashing reduced': [497.155, 184.2509185187417],\n",
       " 'transgene expression increased': [114.59, 89.566243082983],\n",
       " 'transgene expression reduced': [169.415, 145.34246721106672],\n",
       " 'transgene expression reduced male': [453.08, 129.97824279470777],\n",
       " 'transgene expression undetectable': [186.48, 80.29177791031907],\n",
       " 'transgene expression undetectable cell specific': [434.325,\n",
       "  132.59822538405257],\n",
       " 'transgene expression variant': [219.73, 155.73373783480574],\n",
       " 'transgene induced cosuppression variant': [279.665, 170.9894522331714],\n",
       " 'transgene subcellular localization variant': [127.61, 119.98590708912444],\n",
       " 'translation variant': [265.35, 102.9572605501914],\n",
       " 'transposon silencing variant': [358.21, 149.2987136582228],\n",
       " 'tumorous germline': [256.56, 162.2469919597895],\n",
       " 'tunicamycin hypersensitive': [294.145, 158.03614768463575],\n",
       " 'two fold arrest': [323.04, 83.23868331491074],\n",
       " 'unfertilized oocytes laid': [307.7, 108.55648299387744],\n",
       " 'unfolded protein accumulation': [298.95, 95.16909950188663],\n",
       " 'unfolded protein response variant': [294.87, 102.52176890787635],\n",
       " 'univalent meiotic chromosomes': [497.03, 219.74305244990114],\n",
       " 'uptake of yolk defective': [519.66, 245.36783489284],\n",
       " 'uterus morphology variant': [307.165, 108.92390818823937],\n",
       " 'variable cell division timing': [200.745, 99.80711384966504],\n",
       " 'ventral closure defective': [242.05, 103.41811978565458],\n",
       " 'ventral cord patterning variant': [269.925, 155.53745328698164],\n",
       " 'ventral enclosure variant': [465.205, 216.86957595522708],\n",
       " 'vulva cell fate specification variant': [307.54, 99.84326917724599],\n",
       " 'vulva development variant': [307.71, 102.0821037204857],\n",
       " 'vulva morphology variant': [339.435, 181.22989757487588],\n",
       " 'vulval cell induction increased': [416.495, 125.02715695000026],\n",
       " 'vulval cell induction variant': [279.91, 95.4858727770763],\n",
       " 'vulvaless': [381.1, 225.85203120627452]}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iterations_200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_200_iterations = pd.DataFrame.from_dict(iterations_200, orient='index', columns = ['mean', 'sd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lysosome morphology variant</th>\n",
       "      <td>377.530</td>\n",
       "      <td>133.995892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>body elongation defective</th>\n",
       "      <td>275.525</td>\n",
       "      <td>178.061785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hypodermis disorganized</th>\n",
       "      <td>438.780</td>\n",
       "      <td>126.478700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pore forming toxin hypersensitive</th>\n",
       "      <td>218.550</td>\n",
       "      <td>163.641644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transgene expression reduced</th>\n",
       "      <td>169.415</td>\n",
       "      <td>145.342467</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      mean          sd\n",
       "lysosome morphology variant        377.530  133.995892\n",
       "body elongation defective          275.525  178.061785\n",
       "hypodermis disorganized            438.780  126.478700\n",
       "pore forming toxin hypersensitive  218.550  163.641644\n",
       "transgene expression reduced       169.415  145.342467"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_200_iterations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>embryonic lethal</th>\n",
       "      <td>39.100</td>\n",
       "      <td>54.056082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>slow growth</th>\n",
       "      <td>63.860</td>\n",
       "      <td>80.597273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>larval arrest</th>\n",
       "      <td>94.455</td>\n",
       "      <td>84.134642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maternal sterile</th>\n",
       "      <td>111.380</td>\n",
       "      <td>94.745003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sterile progeny</th>\n",
       "      <td>112.545</td>\n",
       "      <td>95.139361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     mean         sd\n",
       "embryonic lethal   39.100  54.056082\n",
       "slow growth        63.860  80.597273\n",
       "larval arrest      94.455  84.134642\n",
       "maternal sterile  111.380  94.745003\n",
       "sterile progeny   112.545  95.139361"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_200_iterations.sort_values(by=['mean']).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_200_iterations.to_csv('output/wormbase_rnai_phenotype_greater4_2018_06_stats.tsv', sep='\\t', header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
