{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import json\n",
    "import re\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import pandas as pd\n",
    "import time\n",
    "import math\n",
    "from glob import glob\n",
    "import uuid\n",
    "\n",
    "import pyarrow.feather as feather\n",
    "from tqdm import tqdm\n",
    "from scipy.stats import rankdata, zscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sn = \"Homo_sapiens\"\n",
    "organism = \"../Serialization/Mammalia/%s\"%sn\n",
    "url = 'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/{}.gene_info.gz'.format(organism)\n",
    "file = '{}.gene_info.tsv'.format(organism)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_save_read(url, file, reader=pd.read_csv, sep='\\t', **kwargs):\n",
    "  ''' Download file from {url}, save it to {file}, and subsequently read it with {reader} using pandas options on {**kwargs}.\n",
    "  '''\n",
    "  if not os.path.exists(file):\n",
    "    if os.path.dirname(file):\n",
    "      os.makedirs(os.path.dirname(file), exist_ok=True)\n",
    "    df = reader(url, sep=sep, index_col=None)\n",
    "    df.to_csv(file, sep=sep, index=False)\n",
    "  return pd.read_csv(file, sep=sep, **kwargs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20220805</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20221009</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20221025</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20220925</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20221009</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID Symbol LocusTag                   Synonyms  \\\n",
       "0     9606       1   A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1     9606       2    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2     9606       3  A2MP1        -                       A2MP   \n",
       "3     9606       9   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4     9606      10   NAT2        -            AAC2|NAT-2|PNAT   \n",
       "\n",
       "                                             dbXrefs chromosome map_location  \\\n",
       "0  MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...         19     19q13.43   \n",
       "1  MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...         12     12p13.31   \n",
       "2  HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...         12     12p13.31   \n",
       "3  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...          8         8p22   \n",
       "4  MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...          8         8p22   \n",
       "\n",
       "                          description    type_of_gene  \\\n",
       "0              alpha-1-B glycoprotein  protein-coding   \n",
       "1               alpha-2-macroglobulin  protein-coding   \n",
       "2  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3               N-acetyltransferase 1  protein-coding   \n",
       "4               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                               A1BG                alpha-1-B glycoprotein   \n",
       "1                                A2M                 alpha-2-macroglobulin   \n",
       "2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n",
       "3                               NAT1                 N-acetyltransferase 1   \n",
       "4                               NAT2                 N-acetyltransferase 2   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...   \n",
       "1                   O  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...   \n",
       "2                   O                  pregnancy-zone protein pseudogene   \n",
       "3                   O  arylamine N-acetyltransferase 1|N-acetyltransf...   \n",
       "4                   O  arylamine N-acetyltransferase 2|N-acetyltransf...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20220805            -  \n",
       "1           20221009            -  \n",
       "2           20221025            -  \n",
       "3           20220925            -  \n",
       "4           20221009            -  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_gene = fetch_save_read(url, file)\n",
    "ncbi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def maybe_split(record):\n",
    "    ''' NCBI Stores Nulls as '-' and lists '|' delimited\n",
    "    '''\n",
    "    if record in {'', '-'}:\n",
    "        return set()\n",
    "    return set(record.split('|'))\n",
    "\n",
    "def supplement_dbXref_prefix_omitted(ids):\n",
    "    ''' NCBI Stores external IDS with Foreign:ID while most datasets just use the ID\n",
    "    '''\n",
    "    for id in ids:\n",
    "        # add original id\n",
    "        yield id\n",
    "        # also add id *without* prefix\n",
    "        if ':' in id:\n",
    "            yield id.split(':', maxsplit=1)[1]\n",
    "\n",
    "ncbi_gene['All_synonyms'] = [\n",
    "    set.union(\n",
    "      maybe_split(gene_info['Symbol']),\n",
    "      maybe_split(gene_info['Symbol_from_nomenclature_authority']),\n",
    "      maybe_split(str(gene_info['GeneID'])),\n",
    "      maybe_split(gene_info['Synonyms']),\n",
    "      maybe_split(gene_info['Other_designations']),\n",
    "      maybe_split(gene_info['LocusTag']),\n",
    "      set(supplement_dbXref_prefix_omitted(maybe_split(gene_info['dbXrefs']))),\n",
    "    )\n",
    "    for _, gene_info in ncbi_gene.iterrows()\n",
    "  ]\n",
    "\n",
    "ncbi_gene['All_synonyms'] = [\n",
    "    set.union(\n",
    "      maybe_split(gene_info['Symbol']),\n",
    "      maybe_split(gene_info['Symbol_from_nomenclature_authority']),\n",
    "      maybe_split(str(gene_info['GeneID'])),\n",
    "      maybe_split(gene_info['Synonyms']),\n",
    "      maybe_split(gene_info['Other_designations']),\n",
    "      maybe_split(gene_info['LocusTag']),\n",
    "      set(supplement_dbXref_prefix_omitted(maybe_split(gene_info['dbXrefs']))),\n",
    "    )\n",
    "    for _, gene_info in ncbi_gene.iterrows()\n",
    "  ]\n",
    "\n",
    "synonyms, gene_id = zip(*{\n",
    "    (synonym, gene_info['Symbol'])\n",
    "    for _, gene_info in ncbi_gene.iterrows()\n",
    "    for synonym in gene_info['All_synonyms']\n",
    "  })\n",
    "ncbi_lookup_syn = pd.Series(gene_id, index=synonyms)\n",
    "symbols, cap, gene_id = zip(*{\n",
    "    (gene_info['Symbol'], gene_info['Symbol'].upper(), gene_info['Symbol'])\n",
    "    for _, gene_info in ncbi_gene.iterrows()\n",
    "  })\n",
    "ncbi_lookup_sym = pd.Series(gene_id, index=symbols)\n",
    "ncbi_lookup_sym_cap = pd.Series(gene_id, index=cap)\n",
    "\n",
    "index_values = ncbi_lookup_syn.index.value_counts()\n",
    "ambiguous = index_values[index_values > 1].index\n",
    "ncbi_lookup_syn_disambiguated = ncbi_lookup_syn[(\n",
    "(ncbi_lookup_syn.index == ncbi_lookup_syn) | (~ncbi_lookup_syn.index.isin(ambiguous))\n",
    ")]\n",
    "all_genes = {}\n",
    "\n",
    "def gene_lookup(gene):\n",
    "    gene_id = ncbi_lookup_sym.to_dict().get(gene)\n",
    "    if gene_id: return str(gene_id)\n",
    "    gene_id = ncbi_lookup_sym_cap.to_dict().get(gene)\n",
    "    if gene_id: return str(gene_id)\n",
    "    return str(ncbi_lookup_syn_disambiguated.to_dict().get(gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wq/7bkxhy8j7ql194c1fbby4gy40000gn/T/ipykernel_2690/2743813533.py:1: DtypeWarning: Columns (9,11,12,23,26) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('data/gwas_catalog_v1.0-associations_e109_r2023-05-20.tsv', sep=\"\\t\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>INTERGENIC</th>\n",
       "      <th>RISK ALLELE FREQUENCY</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2017-08-10</td>\n",
       "      <td>28443625</td>\n",
       "      <td>Justice AE</td>\n",
       "      <td>2017-04-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28443625</td>\n",
       "      <td>Genome-wide meta-analysis of 241,258 adults ac...</td>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>98,173 European ancestry women, 64,373 Europea...</td>\n",
       "      <td>21,496 European ancestry women, 24,385 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7603</td>\n",
       "      <td>1.000000e-28</td>\n",
       "      <td>28.00000</td>\n",
       "      <td>(women)</td>\n",
       "      <td>0.0542</td>\n",
       "      <td>[0.045-0.064] kg/m2 decrease</td>\n",
       "      <td>Affymetrix, Illumina, Perlegen [up to 2800000]...</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2017-08-10</td>\n",
       "      <td>28443625</td>\n",
       "      <td>Justice AE</td>\n",
       "      <td>2017-04-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28443625</td>\n",
       "      <td>Genome-wide meta-analysis of 241,258 adults ac...</td>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>98,173 European ancestry women, 64,373 Europea...</td>\n",
       "      <td>21,496 European ancestry women, 24,385 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3933</td>\n",
       "      <td>2.000000e-06</td>\n",
       "      <td>5.69897</td>\n",
       "      <td>(women)</td>\n",
       "      <td>0.0200</td>\n",
       "      <td>[0.012-0.028] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina, Perlegen [up to 2800000]...</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2017-08-10</td>\n",
       "      <td>28443625</td>\n",
       "      <td>Justice AE</td>\n",
       "      <td>2017-04-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28443625</td>\n",
       "      <td>Genome-wide meta-analysis of 241,258 adults ac...</td>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>98,173 European ancestry women, 64,373 Europea...</td>\n",
       "      <td>21,496 European ancestry women, 24,385 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3933</td>\n",
       "      <td>1.000000e-11</td>\n",
       "      <td>11.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0223</td>\n",
       "      <td>[0.016-0.029] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina, Perlegen [up to 2800000]...</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2017-08-10</td>\n",
       "      <td>28443625</td>\n",
       "      <td>Justice AE</td>\n",
       "      <td>2017-04-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28443625</td>\n",
       "      <td>Genome-wide meta-analysis of 241,258 adults ac...</td>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>98,173 European ancestry women, 64,373 Europea...</td>\n",
       "      <td>21,496 European ancestry women, 24,385 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3933</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>7.00000</td>\n",
       "      <td>(men)</td>\n",
       "      <td>0.0254</td>\n",
       "      <td>[0.016-0.035] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina, Perlegen [up to 2800000]...</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2017-08-10</td>\n",
       "      <td>28443625</td>\n",
       "      <td>Justice AE</td>\n",
       "      <td>2017-04-26</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/28443625</td>\n",
       "      <td>Genome-wide meta-analysis of 241,258 adults ac...</td>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>98,173 European ancestry women, 64,373 Europea...</td>\n",
       "      <td>21,496 European ancestry women, 24,385 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1937</td>\n",
       "      <td>8.000000e-07</td>\n",
       "      <td>6.09691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0231</td>\n",
       "      <td>[0.014-0.032] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina, Perlegen [up to 2800000]...</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID FIRST AUTHOR        DATE     JOURNAL  \\\n",
       "0            2017-08-10  28443625   Justice AE  2017-04-26  Nat Commun   \n",
       "1            2017-08-10  28443625   Justice AE  2017-04-26  Nat Commun   \n",
       "2            2017-08-10  28443625   Justice AE  2017-04-26  Nat Commun   \n",
       "3            2017-08-10  28443625   Justice AE  2017-04-26  Nat Commun   \n",
       "4            2017-08-10  28443625   Justice AE  2017-04-26  Nat Commun   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/28443625   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/28443625   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/28443625   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/28443625   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/28443625   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Genome-wide meta-analysis of 241,258 adults ac...   \n",
       "1  Genome-wide meta-analysis of 241,258 adults ac...   \n",
       "2  Genome-wide meta-analysis of 241,258 adults ac...   \n",
       "3  Genome-wide meta-analysis of 241,258 adults ac...   \n",
       "4  Genome-wide meta-analysis of 241,258 adults ac...   \n",
       "\n",
       "                          DISEASE/TRAIT  \\\n",
       "0  BMI (adjusted for smoking behaviour)   \n",
       "1  BMI (adjusted for smoking behaviour)   \n",
       "2  BMI (adjusted for smoking behaviour)   \n",
       "3  BMI (adjusted for smoking behaviour)   \n",
       "4  BMI (adjusted for smoking behaviour)   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  \\\n",
       "0  98,173 European ancestry women, 64,373 Europea...   \n",
       "1  98,173 European ancestry women, 64,373 Europea...   \n",
       "2  98,173 European ancestry women, 64,373 Europea...   \n",
       "3  98,173 European ancestry women, 64,373 Europea...   \n",
       "4  98,173 European ancestry women, 64,373 Europea...   \n",
       "\n",
       "                             REPLICATION SAMPLE SIZE  ...  \\\n",
       "0  21,496 European ancestry women, 24,385 Europea...  ...   \n",
       "1  21,496 European ancestry women, 24,385 Europea...  ...   \n",
       "2  21,496 European ancestry women, 24,385 Europea...  ...   \n",
       "3  21,496 European ancestry women, 24,385 Europea...  ...   \n",
       "4  21,496 European ancestry women, 24,385 Europea...  ...   \n",
       "\n",
       "                 CONTEXT INTERGENIC RISK ALLELE FREQUENCY       P-VALUE  \\\n",
       "0  upstream_gene_variant        1.0                0.7603  1.000000e-28   \n",
       "1         intron_variant        0.0                0.3933  2.000000e-06   \n",
       "2         intron_variant        0.0                0.3933  1.000000e-11   \n",
       "3         intron_variant        0.0                0.3933  1.000000e-07   \n",
       "4         intron_variant        0.0                0.1937  8.000000e-07   \n",
       "\n",
       "  PVALUE_MLOG P-VALUE (TEXT) OR or BETA                 95% CI (TEXT)  \\\n",
       "0    28.00000        (women)     0.0542  [0.045-0.064] kg/m2 decrease   \n",
       "1     5.69897        (women)     0.0200  [0.012-0.028] kg/m2 increase   \n",
       "2    11.00000            NaN     0.0223  [0.016-0.029] kg/m2 increase   \n",
       "3     7.00000          (men)     0.0254  [0.016-0.035] kg/m2 increase   \n",
       "4     6.09691            NaN     0.0231  [0.014-0.032] kg/m2 increase   \n",
       "\n",
       "                          PLATFORM [SNPS PASSING QC]  CNV  \n",
       "0  Affymetrix, Illumina, Perlegen [up to 2800000]...    N  \n",
       "1  Affymetrix, Illumina, Perlegen [up to 2800000]...    N  \n",
       "2  Affymetrix, Illumina, Perlegen [up to 2800000]...    N  \n",
       "3  Affymetrix, Illumina, Perlegen [up to 2800000]...    N  \n",
       "4  Affymetrix, Illumina, Perlegen [up to 2800000]...    N  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('data/gwas_catalog_v1.0-associations_e109_r2023-05-20.tsv', sep=\"\\t\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = ['DISEASE/TRAIT', \n",
    "       'REPORTED GENE(S)', 'MAPPED_GENE', 'UPSTREAM_GENE_ID',\n",
    "       'DOWNSTREAM_GENE_ID',  'SNP_GENE_IDS',  'P-VALUE', 'PVALUE_MLOG', 'P-VALUE (TEXT)',\n",
    "       ]\n",
    "\n",
    "df = df[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>REPORTED GENE(S)</th>\n",
       "      <th>MAPPED_GENE</th>\n",
       "      <th>UPSTREAM_GENE_ID</th>\n",
       "      <th>DOWNSTREAM_GENE_ID</th>\n",
       "      <th>SNP_GENE_IDS</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>RPS3AP49 - RNU4-17P</td>\n",
       "      <td>ENSG00000242060</td>\n",
       "      <td>ENSG00000202468</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e-28</td>\n",
       "      <td>28.00000</td>\n",
       "      <td>(women)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>MTCH2</td>\n",
       "      <td>MTCH2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000109919</td>\n",
       "      <td>2.000000e-06</td>\n",
       "      <td>5.69897</td>\n",
       "      <td>(women)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>MTCH2</td>\n",
       "      <td>MTCH2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000109919</td>\n",
       "      <td>1.000000e-11</td>\n",
       "      <td>11.00000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>MTCH2</td>\n",
       "      <td>MTCH2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000109919</td>\n",
       "      <td>1.000000e-07</td>\n",
       "      <td>7.00000</td>\n",
       "      <td>(men)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMI (adjusted for smoking behaviour)</td>\n",
       "      <td>MTIF3</td>\n",
       "      <td>MTIF3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000122033</td>\n",
       "      <td>8.000000e-07</td>\n",
       "      <td>6.09691</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          DISEASE/TRAIT REPORTED GENE(S)          MAPPED_GENE  \\\n",
       "0  BMI (adjusted for smoking behaviour)             MC4R  RPS3AP49 - RNU4-17P   \n",
       "1  BMI (adjusted for smoking behaviour)            MTCH2                MTCH2   \n",
       "2  BMI (adjusted for smoking behaviour)            MTCH2                MTCH2   \n",
       "3  BMI (adjusted for smoking behaviour)            MTCH2                MTCH2   \n",
       "4  BMI (adjusted for smoking behaviour)            MTIF3                MTIF3   \n",
       "\n",
       "  UPSTREAM_GENE_ID DOWNSTREAM_GENE_ID     SNP_GENE_IDS       P-VALUE  \\\n",
       "0  ENSG00000242060    ENSG00000202468              NaN  1.000000e-28   \n",
       "1              NaN                NaN  ENSG00000109919  2.000000e-06   \n",
       "2              NaN                NaN  ENSG00000109919  1.000000e-11   \n",
       "3              NaN                NaN  ENSG00000109919  1.000000e-07   \n",
       "4              NaN                NaN  ENSG00000122033  8.000000e-07   \n",
       "\n",
       "   PVALUE_MLOG P-VALUE (TEXT)  \n",
       "0     28.00000        (women)  \n",
       "1      5.69897        (women)  \n",
       "2     11.00000            NaN  \n",
       "3      7.00000          (men)  \n",
       "4      6.09691            NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24929"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(df['DISEASE/TRAIT']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 44813/44813 [00:00<00:00, 1664476.51it/s]\n"
     ]
    }
   ],
   "source": [
    "all_genes = {}\n",
    "for genes in tqdm(df['MAPPED_GENE'].unique()):\n",
    "    if type(genes) == str:\n",
    "        for g in genes.split(\",\"):\n",
    "            gene = g.strip()\n",
    "            if gene not in all_genes:\n",
    "                all_genes[gene] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42757"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 42757/42757 [2:39:36<00:00,  4.46it/s]  \n"
     ]
    }
   ],
   "source": [
    "for gene in tqdm(all_genes):\n",
    "    all_genes[gene] = gene_lookup(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"output/all_genes.json\", 'w') as o:\n",
    "    o.write(json.dumps(all_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "519421it [00:08, 59273.67it/s]\n"
     ]
    }
   ],
   "source": [
    "gmt = {}\n",
    "none_counter = {}\n",
    "for i, row in tqdm(df.iterrows()):\n",
    "    disease = row['DISEASE/TRAIT']\n",
    "    if disease not in gmt:\n",
    "        gmt[disease] = set()\n",
    "    genes = row[\"MAPPED_GENE\"]\n",
    "    if type(genes) == str:\n",
    "        for g in genes.split(\",\"):\n",
    "            gene = all_genes[g.strip()]\n",
    "            if gene != 'None':\n",
    "                gmt[disease].add(gene)\n",
    "            else:\n",
    "                if g.strip() not in none_counter:\n",
    "                    none_counter[g.strip()] = 0\n",
    "                none_counter[g.strip()] += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23858"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(none_counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24929"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gmt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('output/GWAS_Catalog_2023.gmt', 'w') as o:\n",
    "    csv_writer = csv.writer(o, delimiter=\"\\t\")\n",
    "    for k,v in gmt.items():\n",
    "        if len(v) >= 5:\n",
    "            row = [k, ''] + list(v)\n",
    "            csv_writer.writerow(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
