{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>marker_accession_id</th>\n",
       "      <th>marker_symbol</th>\n",
       "      <th>phenotyping_center</th>\n",
       "      <th>colony_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>allele_accession_id</th>\n",
       "      <th>allele_symbol</th>\n",
       "      <th>allele_name</th>\n",
       "      <th>strain_accession_id</th>\n",
       "      <th>...</th>\n",
       "      <th>parameter_name</th>\n",
       "      <th>top_level_mp_term_id</th>\n",
       "      <th>top_level_mp_term_name</th>\n",
       "      <th>mp_term_id</th>\n",
       "      <th>mp_term_name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>percentage_change</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>statistical_method</th>\n",
       "      <th>resource_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1354962</td>\n",
       "      <td>Copb2</td>\n",
       "      <td>BCM</td>\n",
       "      <td>COPBB</td>\n",
       "      <td>male</td>\n",
       "      <td>heterozygote</td>\n",
       "      <td>MGI:6257733</td>\n",
       "      <td>Copb2&lt;em1(IMPC)Bay&gt;</td>\n",
       "      <td>endonuclease-mediated mutation 1, Baylor Colle...</td>\n",
       "      <td>MGI:2159965</td>\n",
       "      <td>...</td>\n",
       "      <td>Lean/Body weight</td>\n",
       "      <td>MP:0005378</td>\n",
       "      <td>growth/size/body region phenotype</td>\n",
       "      <td>MP:0003960</td>\n",
       "      <td>increased lean body mass</td>\n",
       "      <td>2.323142e-06</td>\n",
       "      <td>207.527446680018</td>\n",
       "      <td>2.152058</td>\n",
       "      <td>Linear Mixed Model framework, LME, not includi...</td>\n",
       "      <td>IMPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:1926194</td>\n",
       "      <td>Tnip1</td>\n",
       "      <td>MRC Harwell</td>\n",
       "      <td>H-TNIP1-F09-TM1B</td>\n",
       "      <td>male</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>MGI:6195689</td>\n",
       "      <td>Tnip1&lt;tm1b(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>targeted mutation 1b, Helmholtz Zentrum Muench...</td>\n",
       "      <td>MGI:2164831</td>\n",
       "      <td>...</td>\n",
       "      <td>Red blood cell count</td>\n",
       "      <td>MP:0005397</td>\n",
       "      <td>hematopoietic system phenotype</td>\n",
       "      <td>MP:0002875</td>\n",
       "      <td>decreased erythrocyte cell number</td>\n",
       "      <td>1.493378e-30</td>\n",
       "      <td>-354.836359617433</td>\n",
       "      <td>-6.255841</td>\n",
       "      <td>Linear Mixed Model framework, LME, not includi...</td>\n",
       "      <td>IMPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:106417</td>\n",
       "      <td>Cherp</td>\n",
       "      <td>JAX</td>\n",
       "      <td>JR32021</td>\n",
       "      <td>not_considered</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>MGI:6151421</td>\n",
       "      <td>Cherp&lt;em1(IMPC)J&gt;</td>\n",
       "      <td>endonuclease-mediated mutation 1, Jackson</td>\n",
       "      <td>MGI:3056279</td>\n",
       "      <td>...</td>\n",
       "      <td>Outcome</td>\n",
       "      <td>MP:0010768</td>\n",
       "      <td>mortality/aging</td>\n",
       "      <td>MP:0013292</td>\n",
       "      <td>embryonic lethality prior to organogenesis</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>IMPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1929209</td>\n",
       "      <td>Cldn18</td>\n",
       "      <td>UC Davis</td>\n",
       "      <td>ET9093</td>\n",
       "      <td>female</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>MGI:5569064</td>\n",
       "      <td>Cldn18&lt;tm1.1(KOMP)Vlcg&gt;</td>\n",
       "      <td>targeted mutation 1.1, Velocigene</td>\n",
       "      <td>MGI:2683688</td>\n",
       "      <td>...</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>MP:0005389</td>\n",
       "      <td>reproductive system phenotype</td>\n",
       "      <td>MP:0001120</td>\n",
       "      <td>abnormal uterus morphology</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Supplied as data</td>\n",
       "      <td>IMPC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1354170</td>\n",
       "      <td>Cbarp</td>\n",
       "      <td>BCM</td>\n",
       "      <td>CBRPB</td>\n",
       "      <td>not_considered</td>\n",
       "      <td>homozygote</td>\n",
       "      <td>MGI:6324149</td>\n",
       "      <td>Cbarp&lt;tm2e.1(EUCOMM)Hmgu&gt;</td>\n",
       "      <td>targeted mutation 2e.1, Helmholtz Zentrum Muen...</td>\n",
       "      <td>MGI:2159965</td>\n",
       "      <td>...</td>\n",
       "      <td>Bone Area</td>\n",
       "      <td>MP:0005390</td>\n",
       "      <td>skeleton phenotype</td>\n",
       "      <td>MP:0003795</td>\n",
       "      <td>abnormal bone structure</td>\n",
       "      <td>2.471802e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.135925</td>\n",
       "      <td>Linear Mixed Model framework, LME, including W...</td>\n",
       "      <td>IMPC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  marker_accession_id marker_symbol phenotyping_center         colony_id  \\\n",
       "0         MGI:1354962         Copb2                BCM             COPBB   \n",
       "1         MGI:1926194         Tnip1        MRC Harwell  H-TNIP1-F09-TM1B   \n",
       "2          MGI:106417         Cherp                JAX           JR32021   \n",
       "3         MGI:1929209        Cldn18           UC Davis            ET9093   \n",
       "4         MGI:1354170         Cbarp                BCM             CBRPB   \n",
       "\n",
       "              sex      zygosity allele_accession_id  \\\n",
       "0            male  heterozygote         MGI:6257733   \n",
       "1            male    homozygote         MGI:6195689   \n",
       "2  not_considered    homozygote         MGI:6151421   \n",
       "3          female    homozygote         MGI:5569064   \n",
       "4  not_considered    homozygote         MGI:6324149   \n",
       "\n",
       "               allele_symbol  \\\n",
       "0        Copb2<em1(IMPC)Bay>   \n",
       "1    Tnip1<tm1b(EUCOMM)Hmgu>   \n",
       "2          Cherp<em1(IMPC)J>   \n",
       "3    Cldn18<tm1.1(KOMP)Vlcg>   \n",
       "4  Cbarp<tm2e.1(EUCOMM)Hmgu>   \n",
       "\n",
       "                                         allele_name strain_accession_id  ...  \\\n",
       "0  endonuclease-mediated mutation 1, Baylor Colle...         MGI:2159965  ...   \n",
       "1  targeted mutation 1b, Helmholtz Zentrum Muench...         MGI:2164831  ...   \n",
       "2          endonuclease-mediated mutation 1, Jackson         MGI:3056279  ...   \n",
       "3                  targeted mutation 1.1, Velocigene         MGI:2683688  ...   \n",
       "4  targeted mutation 2e.1, Helmholtz Zentrum Muen...         MGI:2159965  ...   \n",
       "\n",
       "         parameter_name top_level_mp_term_id  \\\n",
       "0      Lean/Body weight           MP:0005378   \n",
       "1  Red blood cell count           MP:0005397   \n",
       "2               Outcome           MP:0010768   \n",
       "3                Uterus           MP:0005389   \n",
       "4             Bone Area           MP:0005390   \n",
       "\n",
       "              top_level_mp_term_name  mp_term_id  \\\n",
       "0  growth/size/body region phenotype  MP:0003960   \n",
       "1     hematopoietic system phenotype  MP:0002875   \n",
       "2                    mortality/aging  MP:0013292   \n",
       "3      reproductive system phenotype  MP:0001120   \n",
       "4                 skeleton phenotype  MP:0003795   \n",
       "\n",
       "                                 mp_term_name       p_value  \\\n",
       "0                    increased lean body mass  2.323142e-06   \n",
       "1           decreased erythrocyte cell number  1.493378e-30   \n",
       "2  embryonic lethality prior to organogenesis  0.000000e+00   \n",
       "3                  abnormal uterus morphology  0.000000e+00   \n",
       "4                     abnormal bone structure  2.471802e-05   \n",
       "\n",
       "   percentage_change effect_size  \\\n",
       "0   207.527446680018    2.152058   \n",
       "1  -354.836359617433   -6.255841   \n",
       "2                NaN    1.000000   \n",
       "3                NaN    1.000000   \n",
       "4                NaN   -2.135925   \n",
       "\n",
       "                                  statistical_method resource_name  \n",
       "0  Linear Mixed Model framework, LME, not includi...          IMPC  \n",
       "1  Linear Mixed Model framework, LME, not includi...          IMPC  \n",
       "2                                   Supplied as data          IMPC  \n",
       "3                                   Supplied as data          IMPC  \n",
       "4  Linear Mixed Model framework, LME, including W...          IMPC  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"data/genotype-phenotype-assertions-ALL.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Terms: 827\n",
      "Gene Coverage: 7314\n",
      "Average genes per terms: 46.89117291414752\n",
      "Max genes per terms: 1868\n",
      "Min genes per terms: 1\n"
     ]
    }
   ],
   "source": [
    "gmt = {}\n",
    "genes = set()\n",
    "for i in df.iterrows():\n",
    "    data = i[1]\n",
    "    mp_term_name = data['mp_term_name']\n",
    "    mp_term_id = data['mp_term_id']\n",
    "    label = \"%s (%s)\\t\"%(mp_term_name, mp_term_id)\n",
    "    if label not in gmt:\n",
    "        gmt[label] = set()\n",
    "    \n",
    "    gene = data['marker_symbol']\n",
    "    gene = gene.upper()\n",
    "    genes.add(gene)\n",
    "    gmt[label].add(gene)\n",
    "\n",
    "print(\"Terms: %s\"%len(gmt))\n",
    "\n",
    "print(\"Gene Coverage: %s\"%len(genes))\n",
    "gene_counts = 0\n",
    "max_genes = 0\n",
    "min_genes = 100\n",
    "for i in gmt.values():\n",
    "    gene_counts += len(i)\n",
    "    if len(i) > max_genes:\n",
    "        max_genes = len(i)\n",
    "    if len(i) < min_genes:\n",
    "        min_genes = len(i)\n",
    "\n",
    "print(\"Average genes per terms: %s\"%(gene_counts/len(gmt)))\n",
    "print(\"Max genes per terms: %s\"%(max_genes))\n",
    "print(\"Min genes per terms: %s\"%(min_genes))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/Users/maayanlab/Documents/codebase/enrichr_utils/gmt/KOMP2.gmt\", \"w\") as o:\n",
    "    for k,v in gmt.items():\n",
    "        line = [k] + list(v) + [\"\\n\"]\n",
    "        o.write(\"\\t\".join(line))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.15 ('venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "43fa70dbc3e0e900c09e1e9e147b0eb772ee319912638d9a00663e86ab5de451"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
