{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from glob import glob\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_source</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AABR07022258.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>405</td>\n",
       "      <td>848</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSRNOG00000055633</td>\n",
       "      <td>1</td>\n",
       "      <td>LOC100910067</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AABR07022620.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>122</td>\n",
       "      <td>427</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSRNOG00000058846</td>\n",
       "      <td>1</td>\n",
       "      <td>AABR07022620.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AABR07022926.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>18</td>\n",
       "      <td>85</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSRNOG00000055717</td>\n",
       "      <td>1</td>\n",
       "      <td>AABR07022926.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AABR07024031.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>18673</td>\n",
       "      <td>58717</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSRNOG00000017648</td>\n",
       "      <td>7</td>\n",
       "      <td>Vom2r8</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AABR07024032.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>17425</td>\n",
       "      <td>17528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSRNOG00000056404</td>\n",
       "      <td>1</td>\n",
       "      <td>RF00026</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seqname   source feature  start    end score strand frame  \\\n",
       "0  AABR07022258.1  ensembl    gene    405    848     .      -     .   \n",
       "1  AABR07022620.1  ensembl    gene    122    427     .      -     .   \n",
       "2  AABR07022926.1  ensembl    gene     18     85     .      +     .   \n",
       "3  AABR07024031.1  ensembl    gene  18673  58717     .      -     .   \n",
       "4  AABR07024032.1  ensembl    gene  17425  17528     .      -     .   \n",
       "\n",
       "              gene_id  gene_version       gene_name gene_source  \\\n",
       "0  ENSRNOG00000055633             1    LOC100910067     ensembl   \n",
       "1  ENSRNOG00000058846             1  AABR07022620.1     ensembl   \n",
       "2  ENSRNOG00000055717             1  AABR07022926.1     ensembl   \n",
       "3  ENSRNOG00000017648             7          Vom2r8     ensembl   \n",
       "4  ENSRNOG00000056404             1         RF00026     ensembl   \n",
       "\n",
       "     gene_biotype  \n",
       "0  protein_coding  \n",
       "1  protein_coding  \n",
       "2           miRNA  \n",
       "3  protein_coding  \n",
       "4           snRNA  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_df = pd.read_csv(\"analysis/transcript-rna-seq/feature-annot/TRNSCRPT_FEATURE_ANNOT.txt\", sep=\"\\t\")\n",
    "feature_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_mapper = {}\n",
    "for i, row in feature_df.iterrows():\n",
    "    gene_id = row[\"gene_id\"]\n",
    "    gene_name = row[\"gene_name\"]\n",
    "    gene_mapper[gene_id] = gene_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sex and Timepoint Comparison\n",
    "* Training group (timepoint) vs sex-matched controls\n",
    "* We filter the dataframe for genes that have an adjusted p-value < 0.05. We then get the up and down differentially expressed genes for each specific tissue, sex, and timepoint group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_ID</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sex</th>\n",
       "      <th>comparison_group</th>\n",
       "      <th>assay</th>\n",
       "      <th>covariates</th>\n",
       "      <th>removed_samples</th>\n",
       "      <th>logFC</th>\n",
       "      <th>logFC_se</th>\n",
       "      <th>shrunk_logFC</th>\n",
       "      <th>shrunk_logFC_se</th>\n",
       "      <th>zscore</th>\n",
       "      <th>p_value</th>\n",
       "      <th>comparison_average_intensity</th>\n",
       "      <th>comparison_average_intensity_se</th>\n",
       "      <th>reference_average_intensity</th>\n",
       "      <th>reference_average_intensity_se</th>\n",
       "      <th>adj_p_value</th>\n",
       "      <th>tissue_abbreviation</th>\n",
       "      <th>selection_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>1w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.106894</td>\n",
       "      <td>0.526792</td>\n",
       "      <td>2.387733e-04</td>\n",
       "      <td>0.024905</td>\n",
       "      <td>0.202916</td>\n",
       "      <td>0.839201</td>\n",
       "      <td>9.466459</td>\n",
       "      <td>1.499481</td>\n",
       "      <td>8.785202</td>\n",
       "      <td>2.113512</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>2w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.009610</td>\n",
       "      <td>0.577451</td>\n",
       "      <td>-2.057732e-05</td>\n",
       "      <td>0.026721</td>\n",
       "      <td>-0.016642</td>\n",
       "      <td>0.986722</td>\n",
       "      <td>8.348006</td>\n",
       "      <td>1.503956</td>\n",
       "      <td>8.785202</td>\n",
       "      <td>2.113512</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>4w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.056960</td>\n",
       "      <td>0.523416</td>\n",
       "      <td>-9.578498e-05</td>\n",
       "      <td>0.021470</td>\n",
       "      <td>-0.108823</td>\n",
       "      <td>0.913343</td>\n",
       "      <td>9.995616</td>\n",
       "      <td>2.264256</td>\n",
       "      <td>8.785202</td>\n",
       "      <td>2.113512</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>8w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.627636</td>\n",
       "      <td>0.503587</td>\n",
       "      <td>2.131641e-02</td>\n",
       "      <td>0.094573</td>\n",
       "      <td>1.246331</td>\n",
       "      <td>0.212643</td>\n",
       "      <td>14.985300</td>\n",
       "      <td>2.621802</td>\n",
       "      <td>8.785202</td>\n",
       "      <td>2.113512</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>female</td>\n",
       "      <td>1w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.198305</td>\n",
       "      <td>0.560630</td>\n",
       "      <td>6.254192e-04</td>\n",
       "      <td>0.031551</td>\n",
       "      <td>0.353718</td>\n",
       "      <td>0.723550</td>\n",
       "      <td>12.607312</td>\n",
       "      <td>1.186143</td>\n",
       "      <td>9.859512</td>\n",
       "      <td>0.574047</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>female</td>\n",
       "      <td>2w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.331535</td>\n",
       "      <td>0.645398</td>\n",
       "      <td>-5.570160e-04</td>\n",
       "      <td>0.026494</td>\n",
       "      <td>-0.513690</td>\n",
       "      <td>0.607468</td>\n",
       "      <td>8.048467</td>\n",
       "      <td>0.590803</td>\n",
       "      <td>9.859512</td>\n",
       "      <td>0.574047</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>female</td>\n",
       "      <td>4w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.481579</td>\n",
       "      <td>6.983907e-09</td>\n",
       "      <td>0.003456</td>\n",
       "      <td>0.000282</td>\n",
       "      <td>0.999775</td>\n",
       "      <td>10.175191</td>\n",
       "      <td>2.228992</td>\n",
       "      <td>9.859512</td>\n",
       "      <td>0.574047</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>female</td>\n",
       "      <td>8w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.058231</td>\n",
       "      <td>0.440509</td>\n",
       "      <td>4.784217e-04</td>\n",
       "      <td>0.039949</td>\n",
       "      <td>0.132191</td>\n",
       "      <td>0.894834</td>\n",
       "      <td>10.417179</td>\n",
       "      <td>1.967804</td>\n",
       "      <td>9.859512</td>\n",
       "      <td>0.574047</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSRNOG00000000008</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>1w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.646087</td>\n",
       "      <td>0.491763</td>\n",
       "      <td>-1.675895e-03</td>\n",
       "      <td>0.025519</td>\n",
       "      <td>-1.313817</td>\n",
       "      <td>0.188908</td>\n",
       "      <td>7.980267</td>\n",
       "      <td>1.389427</td>\n",
       "      <td>12.579924</td>\n",
       "      <td>0.896894</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSRNOG00000000008</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>2w</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>pct_globin,rin,pct_umi_dup,median_5_3_bias</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.600313</td>\n",
       "      <td>0.536082</td>\n",
       "      <td>-1.506895e-03</td>\n",
       "      <td>0.027309</td>\n",
       "      <td>-1.119814</td>\n",
       "      <td>0.262793</td>\n",
       "      <td>6.443191</td>\n",
       "      <td>1.643377</td>\n",
       "      <td>12.579924</td>\n",
       "      <td>0.896894</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           feature_ID      tissue     sex comparison_group  \\\n",
       "0  ENSRNOG00000000001  t62-spleen    male               1w   \n",
       "1  ENSRNOG00000000001  t62-spleen    male               2w   \n",
       "2  ENSRNOG00000000001  t62-spleen    male               4w   \n",
       "3  ENSRNOG00000000001  t62-spleen    male               8w   \n",
       "4  ENSRNOG00000000001  t62-spleen  female               1w   \n",
       "5  ENSRNOG00000000001  t62-spleen  female               2w   \n",
       "6  ENSRNOG00000000001  t62-spleen  female               4w   \n",
       "7  ENSRNOG00000000001  t62-spleen  female               8w   \n",
       "8  ENSRNOG00000000008  t62-spleen    male               1w   \n",
       "9  ENSRNOG00000000008  t62-spleen    male               2w   \n",
       "\n",
       "                assay                                  covariates  \\\n",
       "0  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "1  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "2  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "3  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "4  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "5  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "6  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "7  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "8  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "9  transcript-rna-seq  pct_globin,rin,pct_umi_dup,median_5_3_bias   \n",
       "\n",
       "   removed_samples     logFC  logFC_se  shrunk_logFC  shrunk_logFC_se  \\\n",
       "0              NaN  0.106894  0.526792  2.387733e-04         0.024905   \n",
       "1              NaN -0.009610  0.577451 -2.057732e-05         0.026721   \n",
       "2              NaN -0.056960  0.523416 -9.578498e-05         0.021470   \n",
       "3              NaN  0.627636  0.503587  2.131641e-02         0.094573   \n",
       "4              NaN  0.198305  0.560630  6.254192e-04         0.031551   \n",
       "5              NaN -0.331535  0.645398 -5.570160e-04         0.026494   \n",
       "6              NaN  0.000136  0.481579  6.983907e-09         0.003456   \n",
       "7              NaN  0.058231  0.440509  4.784217e-04         0.039949   \n",
       "8              NaN -0.646087  0.491763 -1.675895e-03         0.025519   \n",
       "9              NaN -0.600313  0.536082 -1.506895e-03         0.027309   \n",
       "\n",
       "     zscore   p_value  comparison_average_intensity  \\\n",
       "0  0.202916  0.839201                      9.466459   \n",
       "1 -0.016642  0.986722                      8.348006   \n",
       "2 -0.108823  0.913343                      9.995616   \n",
       "3  1.246331  0.212643                     14.985300   \n",
       "4  0.353718  0.723550                     12.607312   \n",
       "5 -0.513690  0.607468                      8.048467   \n",
       "6  0.000282  0.999775                     10.175191   \n",
       "7  0.132191  0.894834                     10.417179   \n",
       "8 -1.313817  0.188908                      7.980267   \n",
       "9 -1.119814  0.262793                      6.443191   \n",
       "\n",
       "   comparison_average_intensity_se  reference_average_intensity  \\\n",
       "0                         1.499481                     8.785202   \n",
       "1                         1.503956                     8.785202   \n",
       "2                         2.264256                     8.785202   \n",
       "3                         2.621802                     8.785202   \n",
       "4                         1.186143                     9.859512   \n",
       "5                         0.590803                     9.859512   \n",
       "6                         2.228992                     9.859512   \n",
       "7                         1.967804                     9.859512   \n",
       "8                         1.389427                    12.579924   \n",
       "9                         1.643377                    12.579924   \n",
       "\n",
       "   reference_average_intensity_se  adj_p_value tissue_abbreviation  \\\n",
       "0                        2.113512          1.0              SPLEEN   \n",
       "1                        2.113512          1.0              SPLEEN   \n",
       "2                        2.113512          1.0              SPLEEN   \n",
       "3                        2.113512          1.0              SPLEEN   \n",
       "4                        0.574047          1.0              SPLEEN   \n",
       "5                        0.574047          1.0              SPLEEN   \n",
       "6                        0.574047          1.0              SPLEEN   \n",
       "7                        0.574047          1.0              SPLEEN   \n",
       "8                        0.896894          1.0              SPLEEN   \n",
       "9                        0.896894          1.0              SPLEEN   \n",
       "\n",
       "   selection_fdr  \n",
       "0            1.0  \n",
       "1            1.0  \n",
       "2            1.0  \n",
       "3            1.0  \n",
       "4            1.0  \n",
       "5            1.0  \n",
       "6            1.0  \n",
       "7            1.0  \n",
       "8            1.0  \n",
       "9            1.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"analysis/transcript-rna-seq/dea/pass1b-06_t62-spleen_transcript-rna-seq_timewise-dea-fdr_20211008.txt\"\n",
    "tw = pd.read_csv(filename, sep=\"\\t\")\n",
    "tw.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_ID</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sex</th>\n",
       "      <th>comparison_group</th>\n",
       "      <th>logFC</th>\n",
       "      <th>p_value</th>\n",
       "      <th>adj_p_value</th>\n",
       "      <th>tissue_abbreviation</th>\n",
       "      <th>selection_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>1w</td>\n",
       "      <td>0.106894</td>\n",
       "      <td>0.839201</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>2w</td>\n",
       "      <td>-0.009610</td>\n",
       "      <td>0.986722</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>4w</td>\n",
       "      <td>-0.056960</td>\n",
       "      <td>0.913343</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>male</td>\n",
       "      <td>8w</td>\n",
       "      <td>0.627636</td>\n",
       "      <td>0.212643</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>female</td>\n",
       "      <td>1w</td>\n",
       "      <td>0.198305</td>\n",
       "      <td>0.723550</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SPLEEN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           feature_ID      tissue     sex comparison_group     logFC  \\\n",
       "0  ENSRNOG00000000001  t62-spleen    male               1w  0.106894   \n",
       "1  ENSRNOG00000000001  t62-spleen    male               2w -0.009610   \n",
       "2  ENSRNOG00000000001  t62-spleen    male               4w -0.056960   \n",
       "3  ENSRNOG00000000001  t62-spleen    male               8w  0.627636   \n",
       "4  ENSRNOG00000000001  t62-spleen  female               1w  0.198305   \n",
       "\n",
       "    p_value  adj_p_value tissue_abbreviation  selection_fdr  \n",
       "0  0.839201          1.0              SPLEEN            1.0  \n",
       "1  0.986722          1.0              SPLEEN            1.0  \n",
       "2  0.913343          1.0              SPLEEN            1.0  \n",
       "3  0.212643          1.0              SPLEEN            1.0  \n",
       "4  0.723550          1.0              SPLEEN            1.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tw.head()[['feature_ID', 'tissue', 'sex', 'comparison_group', 'logFC',  'p_value', 'adj_p_value', 'tissue_abbreviation',\n",
    "       'selection_fdr']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "analysis/transcript-rna-seq/dea/pass1b-06_t61-colon_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t56-vastus-lateralis_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t55-gastrocnemius_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t62-spleen_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t58-heart_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t59-kidney_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t66-lung_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t53-cortex_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t60-adrenal_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t68-liver_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t99-vena-cava_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t67-small-intestine_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t69-brown-adipose_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t70-white-adipose_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t52-hippocampus_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t54-hypothalamus_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t63-testes_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t30-blood-rna_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t64-ovaries_transcript-rna-seq_timewise-dea-fdr_20211008.txt\n"
     ]
    }
   ],
   "source": [
    "with open(\"MoTrPAC_timewise_2023.gmt\", \"w\") as o:\n",
    "    csv_writer = csv.writer(o, delimiter=\"\\t\")\n",
    "    for filename in glob(\"analysis/transcript-rna-seq/dea/*timewise*.txt\"):\n",
    "        print(filename)\n",
    "        df = pd.read_csv(filename, sep=\"\\t\")\n",
    "        tissues = df.tissue.unique()\n",
    "        sexes = df.sex.unique()\n",
    "        groups = df.comparison_group.unique()\n",
    "        \n",
    "        for tissue in tissues:\n",
    "            for sex in sexes:\n",
    "                for timepoint in groups:\n",
    "                    label = \"%s_%s_%s\"%(tissue, sex, timepoint)\n",
    "                    up_genes = set()\n",
    "                    down_genes = set()\n",
    "                    d = df[(df.tissue == tissue) & (df.sex == sex) & (df.comparison_group == timepoint) & (df.adj_p_value < 0.05)] \n",
    "                    up = d[d.logFC > 0]\n",
    "                    up_genes = set([gene_mapper[i] for i in up.feature_ID])\n",
    "\n",
    "                    down = d[d.logFC < 0]\n",
    "                    down_genes = set([gene_mapper[i] for i in down.feature_ID])\n",
    "                    if (len(up_genes) >= 5):\n",
    "                        csv_writer.writerow([\n",
    "                            label + \"_up\",\n",
    "                            \"\",\n",
    "                            *list(up_genes)\n",
    "                        ])\n",
    "                    if (len(down_genes) >= 5):\n",
    "                        csv_writer.writerow([\n",
    "                            label + \"_down\",\n",
    "                            \"\",\n",
    "                            *list(down_genes)\n",
    "                        ])\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training Response Significance\n",
    "* Computed genes that changed over the training time course computed using likelihood ratio test. \n",
    "* We filter the dataframe for genes that have an adjusted training p-value < 0.05. These are the genes that significantly responded over the course of the raining time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_ID</th>\n",
       "      <th>assay</th>\n",
       "      <th>tissue</th>\n",
       "      <th>removed_samples</th>\n",
       "      <th>lrt_male</th>\n",
       "      <th>p_value_male</th>\n",
       "      <th>full_model_male</th>\n",
       "      <th>reduced_model_male</th>\n",
       "      <th>lrt_female</th>\n",
       "      <th>p_value_female</th>\n",
       "      <th>full_model_female</th>\n",
       "      <th>reduced_model_female</th>\n",
       "      <th>p_value</th>\n",
       "      <th>lrt</th>\n",
       "      <th>full_model</th>\n",
       "      <th>reduced_model</th>\n",
       "      <th>adj_p_value</th>\n",
       "      <th>tissue_abbreviation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.651426</td>\n",
       "      <td>0.617741</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>1.071593</td>\n",
       "      <td>0.898753</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.881892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000000008</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.211815</td>\n",
       "      <td>0.696867</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.914729</td>\n",
       "      <td>0.922438</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.926875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSRNOG00000000010</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.149430</td>\n",
       "      <td>0.886352</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>1.891004</td>\n",
       "      <td>0.755797</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.938281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSRNOG00000000012</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.889917</td>\n",
       "      <td>0.018189</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>8.933280</td>\n",
       "      <td>0.062789</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.008879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.076572</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSRNOG00000000017</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.377714</td>\n",
       "      <td>0.666658</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>3.039546</td>\n",
       "      <td>0.551229</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.735361</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSRNOG00000000021</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.929890</td>\n",
       "      <td>0.569626</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>6.986178</td>\n",
       "      <td>0.136620</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.276529</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.616488</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSRNOG00000000024</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.319273</td>\n",
       "      <td>0.364518</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>6.980177</td>\n",
       "      <td>0.136939</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.199538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.518020</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSRNOG00000000033</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.408972</td>\n",
       "      <td>0.170617</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>15.018139</td>\n",
       "      <td>0.004664</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.006474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.072793</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSRNOG00000000034</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.942355</td>\n",
       "      <td>0.000807</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>2.166775</td>\n",
       "      <td>0.705116</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.004819</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.049617</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSRNOG00000000036</td>\n",
       "      <td>transcript-rna-seq</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.579231</td>\n",
       "      <td>0.031723</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>2.925815</td>\n",
       "      <td>0.570316</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias+group</td>\n",
       "      <td>~pct_globin+rin+pct_umi_dup+median_5_3_bias</td>\n",
       "      <td>0.090683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.325494</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           feature_ID               assay      tissue  removed_samples  \\\n",
       "0  ENSRNOG00000000001  transcript-rna-seq  t62-spleen              NaN   \n",
       "1  ENSRNOG00000000008  transcript-rna-seq  t62-spleen              NaN   \n",
       "2  ENSRNOG00000000010  transcript-rna-seq  t62-spleen              NaN   \n",
       "3  ENSRNOG00000000012  transcript-rna-seq  t62-spleen              NaN   \n",
       "4  ENSRNOG00000000017  transcript-rna-seq  t62-spleen              NaN   \n",
       "5  ENSRNOG00000000021  transcript-rna-seq  t62-spleen              NaN   \n",
       "6  ENSRNOG00000000024  transcript-rna-seq  t62-spleen              NaN   \n",
       "7  ENSRNOG00000000033  transcript-rna-seq  t62-spleen              NaN   \n",
       "8  ENSRNOG00000000034  transcript-rna-seq  t62-spleen              NaN   \n",
       "9  ENSRNOG00000000036  transcript-rna-seq  t62-spleen              NaN   \n",
       "\n",
       "    lrt_male  p_value_male                                    full_model_male  \\\n",
       "0   2.651426      0.617741  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "1   2.211815      0.696867  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "2   1.149430      0.886352  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "3  11.889917      0.018189  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "4   2.377714      0.666658  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "5   2.929890      0.569626  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "6   4.319273      0.364518  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "7   6.408972      0.170617  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "8  18.942355      0.000807  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "9  10.579231      0.031723  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "\n",
       "                            reduced_model_male  lrt_female  p_value_female  \\\n",
       "0  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    1.071593        0.898753   \n",
       "1  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    0.914729        0.922438   \n",
       "2  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    1.891004        0.755797   \n",
       "3  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    8.933280        0.062789   \n",
       "4  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    3.039546        0.551229   \n",
       "5  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    6.986178        0.136620   \n",
       "6  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    6.980177        0.136939   \n",
       "7  ~pct_globin+rin+pct_umi_dup+median_5_3_bias   15.018139        0.004664   \n",
       "8  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    2.166775        0.705116   \n",
       "9  ~pct_globin+rin+pct_umi_dup+median_5_3_bias    2.925815        0.570316   \n",
       "\n",
       "                                   full_model_female  \\\n",
       "0  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "1  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "2  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "3  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "4  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "5  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "6  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "7  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "8  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "9  ~pct_globin+rin+pct_umi_dup+median_5_3_bias+group   \n",
       "\n",
       "                          reduced_model_female   p_value  lrt  full_model  \\\n",
       "0  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.881892  NaN         NaN   \n",
       "1  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.926875  NaN         NaN   \n",
       "2  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.938281  NaN         NaN   \n",
       "3  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.008879  NaN         NaN   \n",
       "4  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.735361  NaN         NaN   \n",
       "5  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.276529  NaN         NaN   \n",
       "6  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.199538  NaN         NaN   \n",
       "7  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.006474  NaN         NaN   \n",
       "8  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.004819  NaN         NaN   \n",
       "9  ~pct_globin+rin+pct_umi_dup+median_5_3_bias  0.090683  NaN         NaN   \n",
       "\n",
       "   reduced_model  adj_p_value tissue_abbreviation  \n",
       "0            NaN     1.000000              SPLEEN  \n",
       "1            NaN     1.000000              SPLEEN  \n",
       "2            NaN     1.000000              SPLEEN  \n",
       "3            NaN     0.076572              SPLEEN  \n",
       "4            NaN     1.000000              SPLEEN  \n",
       "5            NaN     0.616488              SPLEEN  \n",
       "6            NaN     0.518020              SPLEEN  \n",
       "7            NaN     0.072793              SPLEEN  \n",
       "8            NaN     0.049617              SPLEEN  \n",
       "9            NaN     0.325494              SPLEEN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"analysis/transcript-rna-seq/dea/pass1b-06_t62-spleen_transcript-rna-seq_training-dea-fdr_20211008.txt\"\n",
    "tr = pd.read_csv(filename, sep=\"\\t\")\n",
    "tr.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_ID</th>\n",
       "      <th>tissue</th>\n",
       "      <th>lrt_male</th>\n",
       "      <th>p_value_male</th>\n",
       "      <th>lrt_female</th>\n",
       "      <th>p_value_female</th>\n",
       "      <th>p_value</th>\n",
       "      <th>lrt</th>\n",
       "      <th>adj_p_value</th>\n",
       "      <th>tissue_abbreviation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000000001</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>2.651426</td>\n",
       "      <td>0.617741</td>\n",
       "      <td>1.071593</td>\n",
       "      <td>0.898753</td>\n",
       "      <td>0.881892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000000008</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>2.211815</td>\n",
       "      <td>0.696867</td>\n",
       "      <td>0.914729</td>\n",
       "      <td>0.922438</td>\n",
       "      <td>0.926875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSRNOG00000000010</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>1.149430</td>\n",
       "      <td>0.886352</td>\n",
       "      <td>1.891004</td>\n",
       "      <td>0.755797</td>\n",
       "      <td>0.938281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSRNOG00000000012</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>11.889917</td>\n",
       "      <td>0.018189</td>\n",
       "      <td>8.933280</td>\n",
       "      <td>0.062789</td>\n",
       "      <td>0.008879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.076572</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSRNOG00000000017</td>\n",
       "      <td>t62-spleen</td>\n",
       "      <td>2.377714</td>\n",
       "      <td>0.666658</td>\n",
       "      <td>3.039546</td>\n",
       "      <td>0.551229</td>\n",
       "      <td>0.735361</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>SPLEEN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           feature_ID      tissue   lrt_male  p_value_male  lrt_female  \\\n",
       "0  ENSRNOG00000000001  t62-spleen   2.651426      0.617741    1.071593   \n",
       "1  ENSRNOG00000000008  t62-spleen   2.211815      0.696867    0.914729   \n",
       "2  ENSRNOG00000000010  t62-spleen   1.149430      0.886352    1.891004   \n",
       "3  ENSRNOG00000000012  t62-spleen  11.889917      0.018189    8.933280   \n",
       "4  ENSRNOG00000000017  t62-spleen   2.377714      0.666658    3.039546   \n",
       "\n",
       "   p_value_female   p_value  lrt  adj_p_value tissue_abbreviation  \n",
       "0        0.898753  0.881892  NaN     1.000000              SPLEEN  \n",
       "1        0.922438  0.926875  NaN     1.000000              SPLEEN  \n",
       "2        0.755797  0.938281  NaN     1.000000              SPLEEN  \n",
       "3        0.062789  0.008879  NaN     0.076572              SPLEEN  \n",
       "4        0.551229  0.735361  NaN     1.000000              SPLEEN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tr.head()[['feature_ID','tissue', 'lrt_male',\n",
    "       'p_value_male', 'lrt_female',\n",
    "       'p_value_female', 'p_value', 'lrt', 'adj_p_value',\n",
    "       'tissue_abbreviation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "analysis/transcript-rna-seq/dea/pass1b-06_t62-spleen_transcript-rna-seq_training-dea-fdr_20211008.txt 1099\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t61-colon_transcript-rna-seq_training-dea-fdr_20211008.txt 2515\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t56-vastus-lateralis_transcript-rna-seq_training-dea-fdr_20211008.txt 764\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t55-gastrocnemius_transcript-rna-seq_training-dea-fdr_20211008.txt 565\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t66-lung_transcript-rna-seq_training-dea-fdr_20211008.txt 961\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t59-kidney_transcript-rna-seq_training-dea-fdr_20211008.txt 338\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t58-heart_transcript-rna-seq_training-dea-fdr_20211008.txt 720\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t67-small-intestine_transcript-rna-seq_training-dea-fdr_20211008.txt 746\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t69-brown-adipose_transcript-rna-seq_training-dea-fdr_20211008.txt 1631\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t99-vena-cava_transcript-rna-seq_training-dea-fdr_20211008.txt 90\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t60-adrenal_transcript-rna-seq_training-dea-fdr_20211008.txt 4489\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t68-liver_transcript-rna-seq_training-dea-fdr_20211008.txt 276\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t53-cortex_transcript-rna-seq_training-dea-fdr_20211008.txt 105\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t64-ovaries_transcript-rna-seq_training-dea-fdr_20211008.txt 896\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t30-blood-rna_transcript-rna-seq_training-dea-fdr_20211008.txt 1575\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t52-hippocampus_transcript-rna-seq_training-dea-fdr_20211008.txt 244\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t70-white-adipose_transcript-rna-seq_training-dea-fdr_20211008.txt 1741\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t63-testes_transcript-rna-seq_training-dea-fdr_20211008.txt 107\n",
      "analysis/transcript-rna-seq/dea/pass1b-06_t54-hypothalamus_transcript-rna-seq_training-dea-fdr_20211008.txt 16\n"
     ]
    }
   ],
   "source": [
    "with open(\"MoTrPAC_training_significance_2023.gmt\", \"w\") as o:\n",
    "    csv_writer = csv.writer(o, delimiter=\"\\t\")\n",
    "    for filename in glob(\"analysis/transcript-rna-seq/dea/*training*.txt\"):\n",
    "        df = pd.read_csv(filename, sep=\"\\t\")\n",
    "        tissues = df.tissue.unique()\n",
    "        for tissue in tissues:\n",
    "                up_genes = set()\n",
    "                down_genes = set()\n",
    "                d = df[(df.tissue == tissue) & (df.adj_p_value < 0.05)] \n",
    "                genes = set([gene_mapper[i] for i in d.feature_ID])\n",
    "\n",
    "                if (len(genes) >= 5):\n",
    "                    print(filename, len(genes))\n",
    "                    csv_writer.writerow([\n",
    "                        tissue,\n",
    "                        \"\",\n",
    "                        *list(genes)\n",
    "                    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# s1 = pd.read_csv(\"s1.tsv\", sep=\"\\t\")\n",
    "# s1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# s1 = s1[s1.assay == \"TRNSCRPT\"]\n",
    "# s1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tr[tr.adj_p_value < 0.05].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# feature_id = \"ENSRNOG00000000034\"\n",
    "# s1[(s1.tissue == \"SPLEEN\") & (s1.feature_ID == feature_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tw[tw.feature_ID == feature_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
